{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *De novo* Canu and Miniasm assembly\n",
    "\n",
    "###### NOTE: This notebook has high memory and computational requirements\n",
    "\n",
    "Obtaining the assembled sequence of a complete genome is a complex multi-step task. For *De novo* assembly, the simplest elements of the hierarchy are the reads provided after the sequencing. The next level of hierarchy is the alignment of multiple reads without definite order (i.e. contigs). Finally, the top level of the hierarchy corresponds to the sum of two or more contigs where a (near) complete structure of the genome under study is obtained. Ideally, one expects to obtain a single fragment (contig) for each chromosome or a plasmid that is present in the genome. However, most of the times the assemblies are incomplete, especially when dealing with short reads that can be caused by the process of preparing the material or by the technological limitations. Specifically, when a repeat region is longer than the reads, this will create a single contig in the assembly with multiple connections. ONT provide long reads that can solve this problem albeit the per-base quality is still low (approximately 12-15% error rate)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *De novo* Canu assembly pipeline\n",
    "\n",
    "This notebook relies in the popular Canu pipeline (with Racon and Pilon to polish the assembly result) and Miniasm for *de novo* assembly of ONT reads. \n",
    "\n",
    "[Canu](https://github.com/marbl/canu) is a popular assembler based on the Celera Assembler that can reliably assemble complete microbial genomes and almost complete eukaryotic chromosomes. Canu has three stages: correction, trimming and assembly. Each stage can be executed independently or in series. Each of the three stages begins by identifying the overlaps between all the pairs of input reads, where they count k-mers in the reads, creating an indexed store of the overlays. From the input reads, the correction stage generates corrected reads. The trimming step trims non-compatible bases and detects fork adapters, chimeric sequences and other anomalies. The assembly stage builds an assembly graph and the contigs. Canu handles the repetitions probabilistically, by statistically filtering the repetitively induced overlays and retrospectively inspecting the graph for possible errors, thereby reducing the possibility of selecting a repetitive k-mer for the overlap. In this way, Canu performs multiple rounds of read and overlapping error correction.\n",
    "\n",
    "Canu substantially reduces coverage requirements with a low coverage hierarchical assembly. In this case, it is recommended to polish the assembly with short high-quality reads. Canu results are optimal with long-read coverages above 20x. \n",
    "\n",
    "Canu works with either FASTA or FASTQ files (compressed and uncompressed), but FASTQ format is needed to run some of the steps and complete the full pipeline. The help page is called with the \"canu -h\" command. These are the parameters needed for running Canu with our data:\n",
    "\n",
    "- **-p and -d**: Assembly files prefix and output directory. Both parameters can be the same and output directory doesn't have to exist before execution.\n",
    "\n",
    "- **-genomeSize**: The estimated genome size. In our case, 2.1 mbp so we write '2.1m'. We can put letter g for gbp or k for kbp as well.\n",
    "\n",
    "- **-nanopore-raw**: The path to our reads in FASTQ.\n",
    "\n",
    "Canu auto-detects available resources and will configure job sizes based on the resources and genome size that is being assembled. The following code uses advanced options like 'corMemory' and 'corThreads' to limit the resources in order to make Canu work on a resource-limited environment. These options also prevent Canu to stop the process because of a lack of computation power."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "usage:   canu [-version] [-citation] \\\n",
      "              [-haplotype | -correct | -trim | -assemble | -trim-assemble] \\\n",
      "              [-s <assembly-specifications-file>] \\\n",
      "               -p <assembly-prefix> \\\n",
      "               -d <assembly-directory> \\\n",
      "               genomeSize=<number>[g|m|k] \\\n",
      "              [other-options] \\\n",
      "              [-haplotype{NAME} illumina.fastq.gz] \\\n",
      "              [-pacbio-raw |\n",
      "               -pacbio-corrected |\n",
      "               -nanopore-raw |\n",
      "               -nanopore-corrected] file1 file2 ...\n",
      "\n",
      "example: canu -d run1 -p godzilla genomeSize=1g -nanopore-raw reads/*.fasta.gz \n",
      "\n",
      "\n",
      "  To restrict canu to only a specific stage, use:\n",
      "    -haplotype     - generate haplotype-specific reads\n",
      "    -correct       - generate corrected reads\n",
      "    -trim          - generate trimmed reads\n",
      "    -assemble      - generate an assembly\n",
      "    -trim-assemble - generate trimmed reads and then assemble them\n",
      "\n",
      "  The assembly is computed in the -d <assembly-directory>, with output files named\n",
      "  using the -p <assembly-prefix>.  This directory is created if needed.  It is not\n",
      "  possible to run multiple assemblies in the same directory.\n",
      "\n",
      "  The genome size should be your best guess of the haploid genome size of what is being\n",
      "  assembled.  It is used primarily to estimate coverage in reads, NOT as the desired\n",
      "  assembly size.  Fractional values are allowed: '4.7m' equals '4700k' equals '4700000'\n",
      "\n",
      "  Some common options:\n",
      "    useGrid=string\n",
      "      - Run under grid control (true), locally (false), or set up for grid control\n",
      "        but don't submit any jobs (remote)\n",
      "    rawErrorRate=fraction-error\n",
      "      - The allowed difference in an overlap between two raw uncorrected reads.  For lower\n",
      "        quality reads, use a higher number.  The defaults are 0.300 for PacBio reads and\n",
      "        0.500 for Nanopore reads.\n",
      "    correctedErrorRate=fraction-error\n",
      "      - The allowed difference in an overlap between two corrected reads.  Assemblies of\n",
      "        low coverage or data with biological differences will benefit from a slight increase\n",
      "        in this.  Defaults are 0.045 for PacBio reads and 0.144 for Nanopore reads.\n",
      "    gridOptions=string\n",
      "      - Pass string to the command used to submit jobs to the grid.  Can be used to set\n",
      "        maximum run time limits.  Should NOT be used to set memory limits; Canu will do\n",
      "        that for you.\n",
      "    minReadLength=number\n",
      "      - Ignore reads shorter than 'number' bases long.  Default: 1000.\n",
      "    minOverlapLength=number\n",
      "      - Ignore read-to-read overlaps shorter than 'number' bases long.  Default: 500.\n",
      "  A full list of options can be printed with '-options'.  All options can be supplied in\n",
      "  an optional sepc file with the -s option.\n",
      "\n",
      "  For TrioCanu, haplotypes are specified with the -haplotype{NAME} option, with any\n",
      "  number of haplotype-specific Illumina read files after.  The {NAME} of each haplotype\n",
      "  is free text (but only letters and numbers, please).  For example:\n",
      "    -haplotypeNANNY nanny/*gz\n",
      "    -haplotypeBILLY billy1.fasta.gz billy2.fasta.gz\n",
      "\n",
      "  Reads can be either FASTA or FASTQ format, uncompressed, or compressed with gz, bz2 or xz.\n",
      "  Reads are specified by the technology they were generated with, and any processing performed:\n",
      "    -pacbio-raw         <files>      Reads are straight off the machine.\n",
      "    -pacbio-corrected   <files>      Reads have been corrected.\n",
      "    -nanopore-raw       <files>\n",
      "    -nanopore-corrected <files>\n",
      "\n",
      "Complete documentation at http://canu.readthedocs.org/en/latest/\n",
      "\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "execution_count": 1,
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "canu -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Canu snapshot v1.8 +44 changes (r9254 a50e26a75ffccc529bd944b7adb291e2b6e1c24b)\n",
      "--\n",
      "-- CITATIONS\n",
      "--\n",
      "-- Koren S, Walenz BP, Berlin K, Miller JR, Phillippy AM.\n",
      "-- Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation.\n",
      "-- Genome Res. 2017 May;27(5):722-736.\n",
      "-- http://doi.org/10.1101/gr.215087.116\n",
      "-- \n",
      "-- Koren S, Rhie A, Walenz BP, Dilthey AT, Bickhart DM, Kingan SB, Hiendleder S, Williams JL, Smith TPL, Phillippy AM.\n",
      "-- De novo assembly of haplotype-resolved genomes with trio binning.\n",
      "-- Nat Biotechnol. 2018\n",
      "-- https//doi.org/10.1038/nbt.4277\n",
      "-- \n",
      "-- Read and contig alignments during correction, consensus and GFA building use:\n",
      "--   Šošic M, Šikic M.\n",
      "--   Edlib: a C/C ++ library for fast, exact sequence alignment using edit distance.\n",
      "--   Bioinformatics. 2017 May 1;33(9):1394-1395.\n",
      "--   http://doi.org/10.1093/bioinformatics/btw753\n",
      "-- \n",
      "-- Overlaps are generated using:\n",
      "--   Berlin K, et al.\n",
      "--   Assembling large genomes with single-molecule sequencing and locality-sensitive hashing.\n",
      "--   Nat Biotechnol. 2015 Jun;33(6):623-30.\n",
      "--   http://doi.org/10.1038/nbt.3238\n",
      "-- \n",
      "--   Myers EW, et al.\n",
      "--   A Whole-Genome Assembly of Drosophila.\n",
      "--   Science. 2000 Mar 24;287(5461):2196-204.\n",
      "--   http://doi.org/10.1126/science.287.5461.2196\n",
      "-- \n",
      "-- Corrected read consensus sequences are generated using an algorithm derived from FALCON-sense:\n",
      "--   Chin CS, et al.\n",
      "--   Phased diploid genome assembly with single-molecule real-time sequencing.\n",
      "--   Nat Methods. 2016 Dec;13(12):1050-1054.\n",
      "--   http://doi.org/10.1038/nmeth.4035\n",
      "-- \n",
      "-- Contig consensus sequences are generated using an algorithm derived from pbdagcon:\n",
      "--   Chin CS, et al.\n",
      "--   Nonhybrid, finished microbial genome assemblies from long-read SMRT sequencing data.\n",
      "--   Nat Methods. 2013 Jun;10(6):563-9\n",
      "--   http://doi.org/10.1038/nmeth.2474\n",
      "-- \n",
      "-- CONFIGURE CANU\n",
      "--\n",
      "-- Detected Java(TM) Runtime Environment '1.8.0_181' (from 'java') with -d64 support.\n",
      "-- Detected gnuplot version '4.6 patchlevel 6   ' (from 'gnuplot') and image format 'png'.\n",
      "-- Detected 3 CPUs and 8 gigabytes of memory.\n",
      "-- Limited to 6 gigabytes from maxMemory option.\n",
      "-- No grid engine detected, grid disabled.\n",
      "--\n",
      "--                            (tag)Concurrency\n",
      "--                     (tag)Threads          |\n",
      "--            (tag)Memory         |          |\n",
      "--        (tag)         |         |          |     total usage     algorithm\n",
      "--        -------  ------  --------   --------  -----------------  -----------------------------\n",
      "-- Local: meryl      6 GB    3 CPUs x   1 job      6 GB    3 CPUs  (k-mer counting)\n",
      "-- Local: hap        6 GB    3 CPUs x   1 job      6 GB    3 CPUs  (read-to-haplotype assignment)\n",
      "-- Local: cormhap    6 GB    3 CPUs x   1 job      6 GB    3 CPUs  (overlap detection with mhap)\n",
      "-- Local: obtovl     4 GB    3 CPUs x   1 job      4 GB    3 CPUs  (overlap detection)\n",
      "-- Local: utgovl     4 GB    3 CPUs x   1 job      4 GB    3 CPUs  (overlap detection)\n",
      "-- Local: cor        2 GB    2 CPUs x   1 job      2 GB    2 CPUs  (read correction)\n",
      "-- Local: ovb        4 GB    1 CPU  x   1 job      4 GB    1 CPU   (overlap store bucketizer)\n",
      "-- Local: ovs        6 GB    1 CPU  x   1 job      6 GB    1 CPU   (overlap store sorting)\n",
      "-- Local: red        6 GB    3 CPUs x   1 job      6 GB    3 CPUs  (read error detection)\n",
      "-- Local: oea        4 GB    1 CPU  x   1 job      4 GB    1 CPU   (overlap error adjustment)\n",
      "-- Local: bat        6 GB    3 CPUs x   1 job      6 GB    3 CPUs  (contig construction with bogart)\n",
      "-- Local: cns      --- GB    3 CPUs x   1 job    --- GB    3 CPUs  (consensus)\n",
      "-- Local: gfa        6 GB    3 CPUs x   1 job      6 GB    3 CPUs  (GFA alignment and processing)\n",
      "--\n",
      "-- Found Nanopore uncorrected reads in the input files.\n",
      "--\n",
      "-- Generating assembly 'sample' in '/home/jovyan/notebooks/data/sample/canu_output'\n",
      "--\n",
      "-- Parameters:\n",
      "--\n",
      "--  genomeSize        2100000\n",
      "--\n",
      "--  Overlap Generation Limits:\n",
      "--    corOvlErrorRate 0.3200 ( 32.00%)\n",
      "--    obtOvlErrorRate 0.1200 ( 12.00%)\n",
      "--    utgOvlErrorRate 0.1200 ( 12.00%)\n",
      "--\n",
      "--  Overlap Processing Limits:\n",
      "--    corErrorRate    0.5000 ( 50.00%)\n",
      "--    obtErrorRate    0.1200 ( 12.00%)\n",
      "--    utgErrorRate    0.1200 ( 12.00%)\n",
      "--    cnsErrorRate    0.2000 ( 20.00%)\n",
      "--\n",
      "--\n",
      "-- BEGIN CORRECTION\n",
      "--\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 20:59:16 2018 with 2.245 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/sqStoreCreate \\\n",
      "      -o ./sample.seqStore.BUILDING \\\n",
      "      -minlength 50 \\\n",
      "      ./sample.seqStore.ssi \\\n",
      "    > ./sample.seqStore.BUILDING.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 20:59:16 2018 (fast as lightning) with 2.236 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- In sequence store './sample.seqStore':\n",
      "--   Found 3604 reads.\n",
      "--   Found 34840031 bases (16.59 times coverage).\n",
      "--\n",
      "--   Read length histogram (one '*' equals 30.55 reads):\n",
      "--        0   4999   2139 **********************************************************************\n",
      "--     5000   9999    616 ********************\n",
      "--    10000  14999    283 *********\n",
      "--    15000  19999    164 *****\n",
      "--    20000  24999    101 ***\n",
      "--    25000  29999     66 **\n",
      "--    30000  34999     54 *\n",
      "--    35000  39999     29 \n",
      "--    40000  44999     26 \n",
      "--    45000  49999     18 \n",
      "--    50000  54999     13 \n",
      "--    55000  59999      7 \n",
      "--    60000  64999     15 \n",
      "--    65000  69999      6 \n",
      "--    70000  74999      7 \n",
      "--    75000  79999      2 \n",
      "--    80000  84999      6 \n",
      "--    85000  89999      5 \n",
      "--    90000  94999      4 \n",
      "--    95000  99999      3 \n",
      "--   100000 104999      1 \n",
      "--   105000 109999      4 \n",
      "--   110000 114999      3 \n",
      "--   115000 119999      2 \n",
      "--   120000 124999      2 \n",
      "--   125000 129999      1 \n",
      "--   130000 134999      2 \n",
      "--   135000 139999      1 \n",
      "--   140000 144999      2 \n",
      "--   145000 149999      1 \n",
      "--   150000 154999      1 \n",
      "--   155000 159999      3 \n",
      "--   160000 164999      1 \n",
      "--   165000 169999      0 \n",
      "--   170000 174999      0 \n",
      "--   175000 179999      3 \n",
      "--   180000 184999      1 \n",
      "--   185000 189999      2 \n",
      "--   190000 194999      2 \n",
      "--   195000 199999      0 \n",
      "--   200000 204999      1 \n",
      "--   205000 209999      0 \n",
      "--   210000 214999      0 \n",
      "--   215000 219999      0 \n",
      "--   220000 224999      0 \n",
      "--   225000 229999      0 \n",
      "--   230000 234999      0 \n",
      "--   235000 239999      1 \n",
      "--   240000 244999      0 \n",
      "--   245000 249999      0 \n",
      "--   250000 254999      0 \n",
      "--   255000 259999      1 \n",
      "--   260000 264999      0 \n",
      "--   265000 269999      0 \n",
      "--   270000 274999      0 \n",
      "--   275000 279999      1 \n",
      "--   280000 284999      0 \n",
      "--   285000 289999      1 \n",
      "--   290000 294999      0 \n",
      "--   295000 299999      0 \n",
      "--   300000 304999      0 \n",
      "--   305000 309999      0 \n",
      "--   310000 314999      0 \n",
      "--   315000 319999      0 \n",
      "--   320000 324999      0 \n",
      "--   325000 329999      0 \n",
      "--   330000 334999      0 \n",
      "--   335000 339999      0 \n",
      "--   340000 344999      0 \n",
      "--   345000 349999      1 \n",
      "--   350000 354999      0 \n",
      "--   355000 359999      0 \n",
      "--   360000 364999      0 \n",
      "--   365000 369999      0 \n",
      "--   370000 374999      0 \n",
      "--   375000 379999      1 \n",
      "--   380000 384999      0 \n",
      "--   385000 389999      1 \n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 20:59:16 2018 with 2.236 GB free disk space\n",
      "\n",
      "    cd correction/0-mercounts\n",
      "    ./meryl-configure.sh \\\n",
      "    > ./meryl-configure.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 20:59:17 2018 (one second) with 2.236 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--  segments   memory batches\n",
      "--  -------- -------- -------\n",
      "--        01  0.12 GB       1\n",
      "--\n",
      "--  For 3604 reads with 34840031 bases, limit to 1 batch.\n",
      "--  Will count kmers using 01 jobs, each using 2 GB and 3 threads.\n",
      "--\n",
      "-- Finished stage 'merylConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Tue Nov 27 20:59:17 2018 with 2.236 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/0-mercounts\n",
      "    ./meryl-count.sh 1 > ./meryl-count.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 20:59:21 2018 (4 seconds) with 2.089 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 Kmer counting (meryl) outputs.\n",
      "-- Finished stage 'cor-merylCountCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Tue Nov 27 20:59:21 2018 with 2.089 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/0-mercounts\n",
      "    ./meryl-process.sh 1 > ./meryl-process.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 20:59:34 2018 (13 seconds) with 2.223 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "--  16-mers                                                                                           Fraction\n",
      "--    Occurrences   NumMers                                                                         Unique Total\n",
      "--       1-     1         0                                                                        0.0000 0.0000\n",
      "--       2-     2   1748994 ********************************************************************** 0.7725 0.5697\n",
      "--       3-     4    393824 ***************                                                        0.9089 0.7206\n",
      "--       5-     7     60200 **                                                                     0.9607 0.8022\n",
      "--       8-    11     25634 *                                                                      0.9766 0.8419\n",
      "--      12-    16     19377                                                                        0.9863 0.8787\n",
      "--      17-    22     11297                                                                        0.9942 0.9219\n",
      "--      23-    29      2275                                                                        0.9982 0.9508\n",
      "--      30-    37       719                                                                        0.9989 0.9581\n",
      "--      38-    46       450                                                                        0.9992 0.9618\n",
      "--      47-    56       280                                                                        0.9994 0.9648\n",
      "--      57-    67       227                                                                        0.9995 0.9670\n",
      "--      68-    79       160                                                                        0.9996 0.9693\n",
      "--      80-    92       147                                                                        0.9997 0.9712\n",
      "--      93-   106       106                                                                        0.9998 0.9733\n",
      "--     107-   121        87                                                                        0.9998 0.9749\n",
      "--     122-   137        66                                                                        0.9998 0.9765\n",
      "--     138-   154        69                                                                        0.9999 0.9780\n",
      "--     155-   172        43                                                                        0.9999 0.9795\n",
      "--     173-   191        24                                                                        0.9999 0.9806\n",
      "--     192-   211        20                                                                        0.9999 0.9813\n",
      "--     212-   232        20                                                                        0.9999 0.9820\n",
      "--     233-   254         7                                                                        0.9999 0.9827\n",
      "--     255-   277         7                                                                        0.9999 0.9830\n",
      "--     278-   301         7                                                                        0.9999 0.9833\n",
      "--     302-   326        14                                                                        0.9999 0.9837\n",
      "--     327-   352        14                                                                        1.0000 0.9844\n",
      "--     353-   379        11                                                                        1.0000 0.9852\n",
      "--     380-   407         8                                                                        1.0000 0.9858\n",
      "--     408-   436         9                                                                        1.0000 0.9863\n",
      "--     437-   466         9                                                                        1.0000 0.9869\n",
      "--     467-   497         2                                                                        1.0000 0.9876\n",
      "--     498-   529        12                                                                        1.0000 0.9877\n",
      "--     530-   562         5                                                                        1.0000 0.9888\n",
      "--     563-   596         7                                                                        1.0000 0.9892\n",
      "--     597-   631         5                                                                        1.0000 0.9899\n",
      "--     632-   667         5                                                                        1.0000 0.9904\n",
      "--     668-   704         2                                                                        1.0000 0.9909\n",
      "--     705-   742         1                                                                        1.0000 0.9911\n",
      "--     743-   781         3                                                                        1.0000 0.9913\n",
      "--     782-   821         1                                                                        1.0000 0.9916\n",
      "--\n",
      "--           0 (max occurrences)\n",
      "--     6140445 (total mers, non-unique)\n",
      "--     2264160 (distinct mers, non-unique)\n",
      "--           0 (unique mers)\n",
      "-- Finished stage 'meryl-process', reset canuIteration.\n",
      "--\n",
      "-- OVERLAPPER (mhap) (correction)\n",
      "--\n",
      "-- Set corMhapSensitivity=high based on read coverage of 16.\n",
      "--\n",
      "-- PARAMETERS: hashes=768, minMatches=2, threshold=0.78\n",
      "--\n",
      "-- Given 5.4 GB, can fit 8100 reads per block.\n",
      "-- For 2 blocks, set stride to 2 blocks.\n",
      "-- Logging partitioning to 'correction/1-overlapper/partitioning.log'.\n",
      "-- Configured 1 mhap precompute jobs.\n",
      "-- Configured 1 mhap overlap jobs.\n",
      "-- Finished stage 'cor-mhapConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'cormhap' concurrent execution on Tue Nov 27 20:59:34 2018 with 2.223 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/1-overlapper\n",
      "    ./precompute.sh 1 > ./precompute.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:26 2018 (352 seconds) with 2.135 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- All 1 mhap precompute jobs finished successfully.\n",
      "-- Finished stage 'cor-mhapPrecomputeCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'cormhap' concurrent execution on Tue Nov 27 21:05:27 2018 with 2.135 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/1-overlapper\n",
      "    ./mhap.sh 1 > ./mhap.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:35 2018 (8 seconds) with 2.132 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 mhap overlap output files.\n",
      "-- Finished stage 'cor-mhapCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:05:35 2018 with 2.132 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreConfig \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -M 4-6 \\\n",
      "     -L ./1-overlapper/ovljob.files \\\n",
      "     -create ./sample.ovlStore.config \\\n",
      "     > ./sample.ovlStore.config.txt \\\n",
      "    2> ./sample.ovlStore.config.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:35 2018 (fast as lightning) with 2.132 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Creating overlap store correction/sample.ovlStore using:\n",
      "--      1 bucket\n",
      "--      2 slices\n",
      "--        using at most 1 GB memory each\n",
      "-- Finished stage 'cor-overlapStoreConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'ovB' concurrent execution on Tue Nov 27 21:05:35 2018 with 2.132 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/sample.ovlStore.BUILDING\n",
      "    ./scripts/1-bucketize.sh 1 > ./logs/1-bucketize.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:35 2018 (fast as lightning) with 2.127 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Overlap store bucketizer finished.\n",
      "-- Finished stage 'cor-overlapStoreBucketizerCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'ovS' concurrent execution on Tue Nov 27 21:05:35 2018 with 2.127 GB free disk space (2 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/sample.ovlStore.BUILDING\n",
      "    ./scripts/2-sort.sh 1 > ./logs/2-sort.000001.out 2>&1\n",
      "    ./scripts/2-sort.sh 2 > ./logs/2-sort.000002.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:36 2018 (one second) with 2.1 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Overlap store sorter finished.\n",
      "-- Finished stage 'cor-overlapStoreSorterCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:05:36 2018 with 2.1 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreIndexer \\\n",
      "      -O  ./sample.ovlStore.BUILDING \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -C  ./sample.ovlStore.config \\\n",
      "      -delete \\\n",
      "    > ./sample.ovlStore.BUILDING.index.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:36 2018 (in the blink of an eye) with 2.108 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Checking store.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:05:36 2018 with 2.108 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreDump \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -O  ./sample.ovlStore \\\n",
      "     -counts \\\n",
      "     > ./sample.ovlStore/counts.dat 2> ./sample.ovlStore/counts.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:36 2018 (lickety-split) with 2.108 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Overlap store 'correction/sample.ovlStore' successfully constructed.\n",
      "-- Found 341246 overlaps for 2358 reads; 1246 reads have no overlaps.\n",
      "--\n",
      "--\n",
      "-- Purged 0.091 GB in 3 overlap output files.\n",
      "-- Finished stage 'cor-createOverlapStore', reset canuIteration.\n",
      "-- Set corMinCoverage=0 based on read coverage of 16.\n",
      "-- Global filter scores will be estimated.\n",
      "-- Computing correction layouts.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:05:36 2018 with 2.2 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/generateCorrectionLayouts \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -O  ./sample.ovlStore \\\n",
      "      -C  ./sample.corStore.WORKING \\\n",
      "      -eC 80 \\\n",
      "    > ./sample.corStore.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:36 2018 (furiously fast) with 2.197 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Finished stage 'cor-buildCorrectionLayoutsConfigure', reset canuIteration.\n",
      "-- Computing correction layouts.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:05:36 2018 with 2.197 GB free disk space\n",
      "\n",
      "    cd correction/2-correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/filterCorrectionLayouts \\\n",
      "      -S  ../../sample.seqStore \\\n",
      "      -C     ../sample.corStore \\\n",
      "      -R      ./sample.readsToCorrect.WORKING \\\n",
      "      -cc 0 \\\n",
      "      -cl 50 \\\n",
      "      -g  2100000 \\\n",
      "      -c  40 \\\n",
      "    > ./sample.readsToCorrect.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:05:36 2018 (fast as lightning) with 2.197 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--                             original      original\n",
      "--                            raw reads     raw reads\n",
      "--   category                w/overlaps  w/o/overlaps\n",
      "--   -------------------- ------------- -------------\n",
      "--   Number of Reads               2320          1284\n",
      "--   Number of Bases           18949600         59604\n",
      "--   Coverage                     9.024         0.028\n",
      "--   Median                        2277             0\n",
      "--   Mean                          8167            46\n",
      "--   N50                          30997          3024\n",
      "--   Minimum                         50             0\n",
      "--   Maximum                     388351          7195\n",
      "--   \n",
      "--                                        --------corrected---------  ----------rescued----------\n",
      "--                             evidence                     expected                     expected\n",
      "--   category                     reads            raw     corrected            raw     corrected\n",
      "--   -------------------- -------------  ------------- -------------  ------------- -------------\n",
      "--   Number of Reads               2358           2319          2319              0             0\n",
      "--   Number of Bases           19009204       18572201      16458008              0             0\n",
      "--   Coverage                     9.052          8.844         7.837          0.000         0.000\n",
      "--   Median                        2249           2270          1519              0             0\n",
      "--   Mean                          8061           8008          7097              0             0\n",
      "--   N50                          30977          29633         32724              0             0\n",
      "--   Minimum                         50             50             4              0             0\n",
      "--   Maximum                     388351         388351        319526              0             0\n",
      "--   \n",
      "--                        --------uncorrected--------\n",
      "--                                           expected\n",
      "--   category                       raw     corrected\n",
      "--   -------------------- ------------- -------------\n",
      "--   Number of Reads               1285          1285\n",
      "--   Number of Bases             437003        320859\n",
      "--   Coverage                     0.208         0.153\n",
      "--   Median                           0             0\n",
      "--   Mean                           340           249\n",
      "--   N50                              0             0\n",
      "--   Minimum                          0             0\n",
      "--   Maximum                     377399        320859\n",
      "--   \n",
      "--   Maximum Memory          4028799956\n",
      "-- Finished stage 'cor-filterCorrectionLayouts', reset canuIteration.\n",
      "-- Using overlaps no worse than 0.5 fraction error for correcting reads (from corErrorRate parameter).\n",
      "-- Finished stage 'cor-generateCorrectedReadsConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'cor' concurrent execution on Tue Nov 27 21:05:36 2018 with 2.197 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/2-correction\n",
      "    ./correctReads.sh 1 > ./correctReads.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:43 2018 (667 seconds) with 2.16 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 read correction output files.\n",
      "-- Finished stage 'cor-generateCorrectedReadsCheck', reset canuIteration.\n",
      "-- Found 1 read correction output files.\n",
      "-- Finished stage 'cor-generateCorrectedReadsCheck', reset canuIteration.\n",
      "--\n",
      "-- Loading corrected reads into corStore and seqStore.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:16:43 2018 with 2.16 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/loadCorrectedReads \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -C ./sample.corStore \\\n",
      "      -L ./2-correction/corjob.files \\\n",
      "    >  ./sample.loadCorrectedReads.log \\\n",
      "    2> ./sample.loadCorrectedReads.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:43 2018 (in the blink of an eye) with 2.151 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- In sequence store './sample.seqStore':\n",
      "--   Found 2318 reads.\n",
      "--   Found 18367583 bases (8.74 times coverage).\n",
      "--\n",
      "--   Read length histogram (one '*' equals 23.44 reads):\n",
      "--        0   4999   1641 **********************************************************************\n",
      "--     5000   9999    323 *************\n",
      "--    10000  14999    135 *****\n",
      "--    15000  19999     68 **\n",
      "--    20000  24999     28 *\n",
      "--    25000  29999     25 *\n",
      "--    30000  34999     20 \n",
      "--    35000  39999      7 \n",
      "--    40000  44999      9 \n",
      "--    45000  49999      5 \n",
      "--    50000  54999      3 \n",
      "--    55000  59999      1 \n",
      "--    60000  64999      5 \n",
      "--    65000  69999      2 \n",
      "--    70000  74999      3 \n",
      "--    75000  79999      2 \n",
      "--    80000  84999      3 \n",
      "--    85000  89999      2 \n",
      "--    90000  94999      0 \n",
      "--    95000  99999      2 \n",
      "--   100000 104999      1 \n",
      "--   105000 109999      3 \n",
      "--   110000 114999      2 \n",
      "--   115000 119999      1 \n",
      "--   120000 124999      1 \n",
      "--   125000 129999      1 \n",
      "--   130000 134999      2 \n",
      "--   135000 139999      1 \n",
      "--   140000 144999      2 \n",
      "--   145000 149999      1 \n",
      "--   150000 154999      0 \n",
      "--   155000 159999      3 \n",
      "--   160000 164999      1 \n",
      "--   165000 169999      0 \n",
      "--   170000 174999      0 \n",
      "--   175000 179999      3 \n",
      "--   180000 184999      1 \n",
      "--   185000 189999      2 \n",
      "--   190000 194999      2 \n",
      "--   195000 199999      0 \n",
      "--   200000 204999      1 \n",
      "--   205000 209999      0 \n",
      "--   210000 214999      0 \n",
      "--   215000 219999      0 \n",
      "--   220000 224999      0 \n",
      "--   225000 229999      0 \n",
      "--   230000 234999      0 \n",
      "--   235000 239999      1 \n",
      "--   240000 244999      0 \n",
      "--   245000 249999      0 \n",
      "--   250000 254999      0 \n",
      "--   255000 259999      1 \n",
      "--   260000 264999      0 \n",
      "--   265000 269999      0 \n",
      "--   270000 274999      0 \n",
      "--   275000 279999      1 \n",
      "--   280000 284999      0 \n",
      "--   285000 289999      1 \n",
      "--   290000 294999      0 \n",
      "--   295000 299999      0 \n",
      "--   300000 304999      0 \n",
      "--   305000 309999      0 \n",
      "--   310000 314999      0 \n",
      "--   315000 319999      0 \n",
      "--   320000 324999      0 \n",
      "--   325000 329999      0 \n",
      "--   330000 334999      0 \n",
      "--   335000 339999      0 \n",
      "--   340000 344999      0 \n",
      "--   345000 349999      1 \n",
      "--   350000 354999      0 \n",
      "--   355000 359999      0 \n",
      "--   360000 364999      0 \n",
      "--   365000 369999      0 \n",
      "--   370000 374999      0 \n",
      "--   375000 379999      0 \n",
      "--   380000 384999      0 \n",
      "--   385000 389999      1 \n",
      "--\n",
      "-- Purging correctReads output after loading into stores.\n",
      "-- Purged 1 .cns outputs.\n",
      "-- Purged 2 .out job log outputs.\n",
      "--\n",
      "-- Purging overlaps used for correction.\n",
      "-- Finished stage 'cor-loadCorrectedReads', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:16:43 2018 with 2.211 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/sqStoreDumpFASTQ \\\n",
      "      -corrected \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -o ./sample.correctedReads.gz \\\n",
      "      -fasta \\\n",
      "      -nolibname \\\n",
      "    > sample.correctedReads.fasta.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:44 2018 (one second) with 2.205 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Corrected reads saved in 'sample.correctedReads.fasta.gz'.\n",
      "-- Finished stage 'cor-dumpCorrectedReads', reset canuIteration.\n",
      "--\n",
      "--\n",
      "-- BEGIN TRIMMING\n",
      "--\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:16:44 2018 with 2.205 GB free disk space\n",
      "\n",
      "    cd trimming/0-mercounts\n",
      "    ./meryl-configure.sh \\\n",
      "    > ./meryl-configure.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:44 2018 (fast as lightning) with 2.205 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--  segments   memory batches\n",
      "--  -------- -------- -------\n",
      "--        01  0.12 GB       1\n",
      "--\n",
      "--  For 2318 reads with 18367583 bases, limit to 1 batch.\n",
      "--  Will count kmers using 01 jobs, each using 2 GB and 3 threads.\n",
      "--\n",
      "-- Finished stage 'merylConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Tue Nov 27 21:16:44 2018 with 2.205 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd trimming/0-mercounts\n",
      "    ./meryl-count.sh 1 > ./meryl-count.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:46 2018 (2 seconds) with 2.109 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 Kmer counting (meryl) outputs.\n",
      "-- Finished stage 'obt-merylCountCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Tue Nov 27 21:16:46 2018 with 2.109 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd trimming/0-mercounts\n",
      "    ./meryl-process.sh 1 > ./meryl-process.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:59 2018 (13 seconds) with 2.2 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "--  22-mers                                                                                           Fraction\n",
      "--    Occurrences   NumMers                                                                         Unique Total\n",
      "--       1-     1         0                                                                        0.0000 0.0000\n",
      "--       2-     2    283705 ***************************************************************        0.3887 0.1578\n",
      "--       3-     4    312953 ********************************************************************** 0.6720 0.3304\n",
      "--       5-     7     75805 ****************                                                       0.8802 0.5122\n",
      "--       8-    11      7719 *                                                                      0.9255 0.5717\n",
      "--      12-    16      6043 *                                                                      0.9334 0.5883\n",
      "--      17-    22      9453 **                                                                     0.9423 0.6158\n",
      "--      23-    29      7525 *                                                                      0.9543 0.6645\n",
      "--      30-    37     21246 ****                                                                   0.9647 0.7214\n",
      "--      38-    46      4987 *                                                                      0.9956 0.9411\n",
      "--      47-    56       112                                                                        0.9993 0.9706\n",
      "--      57-    67        86                                                                        0.9994 0.9722\n",
      "--      68-    79        91                                                                        0.9996 0.9737\n",
      "--      80-    92        59                                                                        0.9997 0.9755\n",
      "--      93-   106        21                                                                        0.9998 0.9769\n",
      "--     107-   121        12                                                                        0.9998 0.9774\n",
      "--     122-   137        22                                                                        0.9998 0.9778\n",
      "--     138-   154         7                                                                        0.9998 0.9786\n",
      "--     155-   172         7                                                                        0.9998 0.9789\n",
      "--     173-   191         5                                                                        0.9998 0.9792\n",
      "--     192-   211         0                                                                        0.0000 0.0000\n",
      "--     212-   232         4                                                                        0.9999 0.9794\n",
      "--     233-   254         2                                                                        0.9999 0.9797\n",
      "--     255-   277         0                                                                        0.0000 0.0000\n",
      "--     278-   301         6                                                                        0.9999 0.9798\n",
      "--     302-   326         8                                                                        0.9999 0.9803\n",
      "--     327-   352        11                                                                        0.9999 0.9810\n",
      "--     353-   379         8                                                                        0.9999 0.9821\n",
      "--     380-   407        10                                                                        0.9999 0.9829\n",
      "--     408-   436         5                                                                        0.9999 0.9840\n",
      "--     437-   466         4                                                                        0.9999 0.9846\n",
      "--     467-   497         6                                                                        0.9999 0.9851\n",
      "--     498-   529         5                                                                        0.9999 0.9859\n",
      "--     530-   562         2                                                                        0.9999 0.9866\n",
      "--     563-   596         0                                                                        0.0000 0.0000\n",
      "--     597-   631         5                                                                        0.9999 0.9870\n",
      "--     632-   667        12                                                                        1.0000 0.9878\n",
      "--     668-   704         2                                                                        1.0000 0.9900\n",
      "--     705-   742         4                                                                        1.0000 0.9904\n",
      "--     743-   781         5                                                                        1.0000 0.9912\n",
      "--\n",
      "--           0 (max occurrences)\n",
      "--     3595515 (total mers, non-unique)\n",
      "--      729967 (distinct mers, non-unique)\n",
      "--           0 (unique mers)\n",
      "-- Finished stage 'meryl-process', reset canuIteration.\n",
      "--\n",
      "-- OVERLAPPER (normal) (trimming) erate=0.12\n",
      "--\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:16:59 2018 with 2.199 GB free disk space\n",
      "\n",
      "    cd trimming/1-overlapper\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/overlapInCorePartition \\\n",
      "     -S  ../../sample.seqStore \\\n",
      "     -hl 80000000 \\\n",
      "     -rl 1000000000 \\\n",
      "     -ol 50 \\\n",
      "     -o  ./sample.partition \\\n",
      "    > ./sample.partition.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:16:59 2018 (in the blink of an eye) with 2.199 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Configured 1 overlapInCore jobs.\n",
      "-- Finished stage 'obt-overlapConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'obtovl' concurrent execution on Tue Nov 27 21:16:59 2018 with 2.199 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd trimming/1-overlapper\n",
      "    ./overlap.sh 1 > ./overlap.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:10 2018 (71 seconds) with 2.197 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 overlapInCore output files.\n",
      "--\n",
      "-- overlapInCore compute 'trimming/1-overlapper':\n",
      "--   kmer hits\n",
      "--     with no overlap            62784       62784 +- 0\n",
      "--     with an overlap           133810      133810 +- 0\n",
      "--\n",
      "--   overlaps                    133801      133801 +- 0\n",
      "--     contained                      0           0 +- 0\n",
      "--     dovetail                       0           0 +- 0\n",
      "--\n",
      "--   overlaps rejected\n",
      "--     multiple per pair              0           0 +- 0\n",
      "--     bad short window               0           0 +- 0\n",
      "--     bad long window                0           0 +- 0\n",
      "-- Finished stage 'obt-overlapCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:10 2018 with 2.197 GB free disk space\n",
      "\n",
      "    cd trimming\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreConfig \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -M 1 \\\n",
      "     -L ./1-overlapper/ovljob.files \\\n",
      "     -create ./sample.ovlStore.config \\\n",
      "     > ./sample.ovlStore.config.txt \\\n",
      "    2> ./sample.ovlStore.config.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:10 2018 (lickety-split) with 2.197 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Creating overlap store trimming/sample.ovlStore using:\n",
      "--      1 bucket\n",
      "--      2 slices\n",
      "--        using at most 1 GB memory each\n",
      "-- Finished stage 'obt-overlapStoreConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'ovB' concurrent execution on Tue Nov 27 21:18:10 2018 with 2.197 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd trimming/sample.ovlStore.BUILDING\n",
      "    ./scripts/1-bucketize.sh 1 > ./logs/1-bucketize.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:10 2018 (furiously fast) with 2.194 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Overlap store bucketizer finished.\n",
      "-- Finished stage 'obt-overlapStoreBucketizerCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'ovS' concurrent execution on Tue Nov 27 21:18:10 2018 with 2.194 GB free disk space (2 processes; 1 concurrently)\n",
      "\n",
      "    cd trimming/sample.ovlStore.BUILDING\n",
      "    ./scripts/2-sort.sh 1 > ./logs/2-sort.000001.out 2>&1\n",
      "    ./scripts/2-sort.sh 2 > ./logs/2-sort.000002.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (one second) with 2.179 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Overlap store sorter finished.\n",
      "-- Finished stage 'obt-overlapStoreSorterCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.179 GB free disk space\n",
      "\n",
      "    cd trimming\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreIndexer \\\n",
      "      -O  ./sample.ovlStore.BUILDING \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -C  ./sample.ovlStore.config \\\n",
      "      -delete \\\n",
      "    > ./sample.ovlStore.BUILDING.index.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (lickety-split) with 2.184 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Checking store.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.184 GB free disk space\n",
      "\n",
      "    cd trimming\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreDump \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -O  ./sample.ovlStore \\\n",
      "     -counts \\\n",
      "     > ./sample.ovlStore/counts.dat 2> ./sample.ovlStore/counts.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (fast as lightning) with 2.184 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Overlap store 'trimming/sample.ovlStore' successfully constructed.\n",
      "-- Found 267620 overlaps for 1525 reads; 2079 reads have no overlaps.\n",
      "--\n",
      "--\n",
      "-- Purged 0.001 GB in 3 overlap output files.\n",
      "-- Finished stage 'obt-createOverlapStore', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.186 GB free disk space\n",
      "\n",
      "    cd trimming/3-overlapbasedtrimming\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/trimReads \\\n",
      "      -S  ../../sample.seqStore \\\n",
      "      -O  ../sample.ovlStore \\\n",
      "      -Co ./sample.1.trimReads.clear \\\n",
      "      -e  0.12 \\\n",
      "      -minlength 50 \\\n",
      "      -ol 1 \\\n",
      "      -oc 1 \\\n",
      "      -o  ./sample.1.trimReads \\\n",
      "    >     ./sample.1.trimReads.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (in the blink of an eye) with 2.186 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--  PARAMETERS:\n",
      "--  ----------\n",
      "--       50    (reads trimmed below this many bases are deleted)\n",
      "--   0.1200    (use overlaps at or below this fraction error)\n",
      "--        1    (break region if overlap is less than this long, for 'largest covered' algorithm)\n",
      "--        1    (break region if overlap coverage is less than this many read, for 'largest covered' algorithm)\n",
      "--  \n",
      "--  INPUT READS:\n",
      "--  -----------\n",
      "--    3604 reads     18367583 bases (reads processed)\n",
      "--       0 reads            0 bases (reads not processed, previously deleted)\n",
      "--       0 reads            0 bases (reads not processed, in a library where trimming isn't allowed)\n",
      "--  \n",
      "--  OUTPUT READS:\n",
      "--  ------------\n",
      "--    1505 reads      3997099 bases (trimmed reads output)\n",
      "--      20 reads        86454 bases (reads with no change, kept as is)\n",
      "--    2079 reads      8730597 bases (reads with no overlaps, deleted)\n",
      "--       0 reads            0 bases (reads with short trimmed length, deleted)\n",
      "--  \n",
      "--  TRIMMING DETAILS:\n",
      "--  ----------------\n",
      "--    1238 reads      2710668 bases (bases trimmed from the 5' end of a read)\n",
      "--    1436 reads      2842765 bases (bases trimmed from the 3' end of a read)\n",
      "-- Finished stage 'obt-trimReads', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.186 GB free disk space\n",
      "\n",
      "    cd trimming/3-overlapbasedtrimming\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/splitReads \\\n",
      "      -S  ../../sample.seqStore \\\n",
      "      -O  ../sample.ovlStore \\\n",
      "      -Ci ./sample.1.trimReads.clear \\\n",
      "      -Co ./sample.2.splitReads.clear \\\n",
      "      -e  0.12 \\\n",
      "      -minlength 50 \\\n",
      "      -o  ./sample.2.splitReads \\\n",
      "    >     ./sample.2.splitReads.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (in the blink of an eye) with 2.186 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--  PARAMETERS:\n",
      "--  ----------\n",
      "--       50    (reads trimmed below this many bases are deleted)\n",
      "--   0.1200    (use overlaps at or below this fraction error)\n",
      "--  INPUT READS:\n",
      "--  -----------\n",
      "--    1525 reads      9636986 bases (reads processed)\n",
      "--    2079 reads      8730597 bases (reads not processed, previously deleted)\n",
      "--       0 reads            0 bases (reads not processed, in a library where trimming isn't allowed)\n",
      "--  \n",
      "--  PROCESSED:\n",
      "--  --------\n",
      "--       0 reads            0 bases (no overlaps)\n",
      "--     156 reads       922509 bases (no coverage after adjusting for trimming done already)\n",
      "--       0 reads            0 bases (processed for chimera)\n",
      "--       0 reads            0 bases (processed for spur)\n",
      "--    1369 reads      8714477 bases (processed for subreads)\n",
      "--  \n",
      "--  READS WITH SIGNALS:\n",
      "--  ------------------\n",
      "--       0 reads            0 signals (number of 5' spur signal)\n",
      "--       0 reads            0 signals (number of 3' spur signal)\n",
      "--       0 reads            0 signals (number of chimera signal)\n",
      "--       0 reads            0 signals (number of subread signal)\n",
      "--  \n",
      "--  SIGNALS:\n",
      "--  -------\n",
      "--       0 reads            0 bases (size of 5' spur signal)\n",
      "--       0 reads            0 bases (size of 3' spur signal)\n",
      "--       0 reads            0 bases (size of chimera signal)\n",
      "--       0 reads            0 bases (size of subread signal)\n",
      "--  \n",
      "--  TRIMMING:\n",
      "--  --------\n",
      "--       0 reads            0 bases (trimmed from the 5' end of the read)\n",
      "--       0 reads            0 bases (trimmed from the 3' end of the read)\n",
      "-- Finished stage 'obt-splitReads', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.186 GB free disk space\n",
      "\n",
      "    cd trimming/3-overlapbasedtrimming\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/loadTrimmedReads \\\n",
      "      -S ../../sample.seqStore \\\n",
      "      -c ./sample.2.splitReads.clear \\\n",
      "    > ./sample.loadtrimmedReads.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (in the blink of an eye) with 2.186 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- In sequence store './sample.seqStore':\n",
      "--   Found 1525 reads.\n",
      "--   Found 4083553 bases (1.94 times coverage).\n",
      "--\n",
      "--   Read length histogram (one '*' equals 10.77 reads):\n",
      "--        0    999    754 **********************************************************************\n",
      "--     1000   1999    253 ***********************\n",
      "--     2000   2999    147 *************\n",
      "--     3000   3999     78 *******\n",
      "--     4000   4999     53 ****\n",
      "--     5000   5999     46 ****\n",
      "--     6000   6999     37 ***\n",
      "--     7000   7999     15 *\n",
      "--     8000   8999     20 *\n",
      "--     9000   9999     25 **\n",
      "--    10000  10999     14 *\n",
      "--    11000  11999     10 \n",
      "--    12000  12999     10 \n",
      "--    13000  13999     14 *\n",
      "--    14000  14999      8 \n",
      "--    15000  15999      3 \n",
      "--    16000  16999      6 \n",
      "--    17000  17999      2 \n",
      "--    18000  18999      2 \n",
      "--    19000  19999      6 \n",
      "--    20000  20999      3 \n",
      "--    21000  21999      2 \n",
      "--    22000  22999      1 \n",
      "--    23000  23999      0 \n",
      "--    24000  24999      2 \n",
      "--    25000  25999      5 \n",
      "--    26000  26999      1 \n",
      "--    27000  27999      2 \n",
      "--    28000  28999      1 \n",
      "--    29000  29999      0 \n",
      "--    30000  30999      0 \n",
      "--    31000  31999      3 \n",
      "--    32000  32999      0 \n",
      "--    33000  33999      0 \n",
      "--    34000  34999      2 \n",
      "--\n",
      "-- Purging overlaps used for trimming.\n",
      "-- Finished stage 'obt-dumpReads', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.199 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/sqStoreDumpFASTQ \\\n",
      "      -trimmed \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -o ./sample.trimmedReads.gz \\\n",
      "      -fasta \\\n",
      "      -nolibname \\\n",
      "    > ./sample.trimmedReads.fasta.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (like a bat out of hell) with 2.198 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Trimmed reads saved in 'sample.trimmedReads.fasta.gz'.\n",
      "-- Finished stage 'cor-dumpTrimmedReads', reset canuIteration.\n",
      "--\n",
      "--\n",
      "-- BEGIN ASSEMBLY\n",
      "--\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:11 2018 with 2.198 GB free disk space\n",
      "\n",
      "    cd unitigging/0-mercounts\n",
      "    ./meryl-configure.sh \\\n",
      "    > ./meryl-configure.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:11 2018 (in the blink of an eye) with 2.198 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--  segments   memory batches\n",
      "--  -------- -------- -------\n",
      "--        01  0.06 GB       1\n",
      "--\n",
      "--  For 1525 reads with 4083553 bases, limit to 1 batch.\n",
      "--  Will count kmers using 01 jobs, each using 2 GB and 3 threads.\n",
      "--\n",
      "-- Finished stage 'merylConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Tue Nov 27 21:18:11 2018 with 2.198 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/0-mercounts\n",
      "    ./meryl-count.sh 1 > ./meryl-count.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:13 2018 (2 seconds) with 2.186 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 Kmer counting (meryl) outputs.\n",
      "-- Finished stage 'utg-merylCountCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Tue Nov 27 21:18:13 2018 with 2.186 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/0-mercounts\n",
      "    ./meryl-process.sh 1 > ./meryl-process.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:27 2018 (14 seconds) with 2.193 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "--  22-mers                                                                                           Fraction\n",
      "--    Occurrences   NumMers                                                                         Unique Total\n",
      "--       1-     1         0                                                                        0.0000 0.0000\n",
      "--       2-     2    204079 ***********************************************************            0.3655 0.1377\n",
      "--       3-     4    239475 ********************************************************************** 0.6446 0.2955\n",
      "--       5-     7     62118 ******************                                                     0.8626 0.4725\n",
      "--       8-    11      5159 *                                                                      0.9096 0.5296\n",
      "--      12-    16      5388 *                                                                      0.9166 0.5432\n",
      "--      17-    22      8777 **                                                                     0.9272 0.5737\n",
      "--      23-    29      7369 **                                                                     0.9419 0.6292\n",
      "--      30-    37     21327 ******                                                                 0.9553 0.6970\n",
      "--      38-    46      4513 *                                                                      0.9955 0.9616\n",
      "--      47-    56        32                                                                        0.9998 0.9940\n",
      "--      57-    67         2                                                                        0.9999 0.9945\n",
      "--      68-    79         6                                                                        0.9999 0.9945\n",
      "--      80-    92         0                                                                        0.0000 0.0000\n",
      "--      93-   106         1                                                                        0.9999 0.9947\n",
      "--     107-   121         1                                                                        0.9999 0.9947\n",
      "--     122-   137         7                                                                        0.9999 0.9948\n",
      "--     138-   154         1                                                                        0.9999 0.9951\n",
      "--     155-   172         2                                                                        0.9999 0.9952\n",
      "--     173-   191         2                                                                        0.9999 0.9953\n",
      "--     192-   211         3                                                                        0.9999 0.9954\n",
      "--     212-   232         5                                                                        0.9999 0.9957\n",
      "--     233-   254         3                                                                        0.9999 0.9960\n",
      "--     255-   277        21                                                                        0.9999 0.9964\n",
      "--     278-   301        10                                                                        1.0000 0.9981\n",
      "--\n",
      "--           0 (max occurrences)\n",
      "--     2963512 (total mers, non-unique)\n",
      "--      558304 (distinct mers, non-unique)\n",
      "--           0 (unique mers)\n",
      "-- Finished stage 'meryl-process', reset canuIteration.\n",
      "--\n",
      "-- OVERLAPPER (normal) (assembly) erate=0.12\n",
      "--\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:27 2018 with 2.193 GB free disk space\n",
      "\n",
      "    cd unitigging/1-overlapper\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/overlapInCorePartition \\\n",
      "     -S  ../../sample.seqStore \\\n",
      "     -hl 80000000 \\\n",
      "     -rl 1000000000 \\\n",
      "     -ol 50 \\\n",
      "     -o  ./sample.partition \\\n",
      "    > ./sample.partition.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:27 2018 (like a bat out of hell) with 2.193 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Configured 1 overlapInCore jobs.\n",
      "-- Finished stage 'utg-overlapConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'utgovl' concurrent execution on Tue Nov 27 21:18:27 2018 with 2.193 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/1-overlapper\n",
      "    ./overlap.sh 1 > ./overlap.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:44 2018 (17 seconds) with 2.193 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 overlapInCore output files.\n",
      "--\n",
      "-- overlapInCore compute 'unitigging/1-overlapper':\n",
      "--   kmer hits\n",
      "--     with no overlap             7184        7184 +- 0\n",
      "--     with an overlap            18247       18247 +- 0\n",
      "--\n",
      "--   overlaps                     18247       18247 +- 0\n",
      "--     contained                  10266       10266 +- 0\n",
      "--     dovetail                    7981        7981 +- 0\n",
      "--\n",
      "--   overlaps rejected\n",
      "--     multiple per pair              0           0 +- 0\n",
      "--     bad short window               0           0 +- 0\n",
      "--     bad long window                0           0 +- 0\n",
      "-- Finished stage 'utg-overlapCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:44 2018 with 2.193 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreConfig \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -M 1 \\\n",
      "     -L ./1-overlapper/ovljob.files \\\n",
      "     -create ./sample.ovlStore.config \\\n",
      "     > ./sample.ovlStore.config.txt \\\n",
      "    2> ./sample.ovlStore.config.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:44 2018 (lickety-split) with 2.193 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Creating overlap store unitigging/sample.ovlStore using:\n",
      "--      1 bucket\n",
      "--      2 slices\n",
      "--        using at most 1 GB memory each\n",
      "-- Finished stage 'utg-overlapStoreConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'ovB' concurrent execution on Tue Nov 27 21:18:44 2018 with 2.193 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/sample.ovlStore.BUILDING\n",
      "    ./scripts/1-bucketize.sh 1 > ./logs/1-bucketize.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:44 2018 (lickety-split) with 2.192 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Overlap store bucketizer finished.\n",
      "-- Finished stage 'utg-overlapStoreBucketizerCheck', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'ovS' concurrent execution on Tue Nov 27 21:18:44 2018 with 2.192 GB free disk space (2 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/sample.ovlStore.BUILDING\n",
      "    ./scripts/2-sort.sh 1 > ./logs/2-sort.000001.out 2>&1\n",
      "    ./scripts/2-sort.sh 2 > ./logs/2-sort.000002.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:45 2018 (one second) with 2.19 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Overlap store sorter finished.\n",
      "-- Finished stage 'utg-overlapStoreSorterCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:45 2018 with 2.19 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreIndexer \\\n",
      "      -O  ./sample.ovlStore.BUILDING \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -C  ./sample.ovlStore.config \\\n",
      "      -delete \\\n",
      "    > ./sample.ovlStore.BUILDING.index.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:45 2018 (lickety-split) with 2.191 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Checking store.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:45 2018 with 2.191 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreDump \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -O  ./sample.ovlStore \\\n",
      "     -counts \\\n",
      "     > ./sample.ovlStore/counts.dat 2> ./sample.ovlStore/counts.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:45 2018 (fast as lightning) with 2.191 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Overlap store 'unitigging/sample.ovlStore' successfully constructed.\n",
      "-- Found 36494 overlaps for 1391 reads; 2213 reads have no overlaps.\n",
      "--\n",
      "--\n",
      "-- Purged 0 GB in 3 overlap output files.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:45 2018 with 2.191 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreStats \\\n",
      "     -C 1 \\\n",
      "     -S ../sample.seqStore \\\n",
      "     -O  ./sample.ovlStore \\\n",
      "     -o  ./sample.ovlStore \\\n",
      "     > ./sample.ovlStore.summary.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:45 2018 (furiously fast) with 2.191 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "--\n",
      "-- Overlap store 'unitigging/sample.ovlStore' contains:\n",
      "--\n",
      "--   category            reads     %          read length        feature size or coverage  analysis\n",
      "--   ----------------  -------  -------  ----------------------  ------------------------  --------------------\n",
      "--   middle-missing         23    1.51     1779.74 +- 2262.22        326.83 +- 513.75     (bad trimming)\n",
      "--   middle-hump            73    4.79     3070.22 +- 4336.97        138.41 +- 343.01     (bad trimming)\n",
      "--   no-5-prime             46    3.02     4615.09 +- 5701.23        100.28 +- 294.14     (bad trimming)\n",
      "--   no-3-prime             65    4.26     3637.86 +- 4195.84        190.65 +- 318.99     (bad trimming)\n",
      "--   \n",
      "--   low-coverage            0    0.00        0.00 +- 0.00             0.00 +- 0.00       (easy to assemble, potential for lower quality consensus)\n",
      "--   unique                 91    5.97      538.63 +- 673.32           1.00 +- 0.00       (easy to assemble, perfect, yay)\n",
      "--   repeat-cont           798   52.33     2827.94 +- 4482.75         21.87 +- 15.74      (potential for consensus errors, no impact on assembly)\n",
      "--   repeat-dove            26    1.70    10385.81 +- 11448.19        17.45 +- 14.79      (hard to assemble, likely won't assemble correctly or even at all)\n",
      "--   \n",
      "--   span-repeat            47    3.08     1885.53 +- 1429.47       1475.34 +- 1367.28    (read spans a large repeat, usually easy to assemble)\n",
      "--   uniq-repeat-cont      170   11.15     2275.69 +- 2732.62                             (should be uniquely placed, low potential for consensus errors, no impact on assembly)\n",
      "--   uniq-repeat-dove       14    0.92     4587.14 +- 4861.51                             (will end contigs, potential to misassemble)\n",
      "--   uniq-anchor            38    2.49     5786.63 +- 6977.12       1274.58 +- 2465.33    (repeat read, with unique section, probable bad read)\n",
      "-- Finished stage 'utg-createOverlapStore', reset canuIteration.\n",
      "--\n",
      "-- Loading read lengths.\n",
      "-- Loading number of overlaps per read.\n",
      "--\n",
      "-- Configure RED for 6gb memory.\n",
      "--                   Batches of at most (unlimited) reads.\n",
      "--                                      500000000 bases.\n",
      "--                   Expecting evidence of at most 536870912 bases per iteration.\n",
      "--\n",
      "--           Total                                               Reads                 Olaps Evidence\n",
      "--    Job   Memory      Read Range         Reads        Bases   Memory        Olaps   Memory   Memory  (Memory in MB)\n",
      "--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------\n",
      "--      1  3119.20         1-3604           1525      4083553    46.78        36494     0.42  1024.00\n",
      "--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------\n",
      "--                                                    4083553                 36494\n",
      "-- Finished stage 'readErrorDetectionConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'red' concurrent execution on Tue Nov 27 21:18:45 2018 with 2.191 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/3-overlapErrorAdjustment\n",
      "    ./red.sh 1 > ./red.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:47 2018 (2 seconds) with 2.191 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 read error detection output files.\n",
      "-- Finished stage 'readErrorDetectionCheck', reset canuIteration.\n",
      "--\n",
      "-- Loading read lengths.\n",
      "-- Loading number of overlaps per read.\n",
      "--\n",
      "-- Configure OEA for 4gb memory.\n",
      "--                   Batches of at most (unlimited) reads.\n",
      "--                                      300000000 bases.\n",
      "--\n",
      "--           Total                                               Reads                 Olaps  Adjusts\n",
      "--    Job   Memory      Read Range         Reads        Bases   Memory        Olaps   Memory   Memory  (Memory in MB)\n",
      "--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------\n",
      "--      1  2381.43         1-3604           1525      4083553    11.94        36494     1.11   320.37\n",
      "--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------\n",
      "--                                                    4083553                 36494\n",
      "-- Finished stage 'overlapErrorAdjustmentConfigure', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'oea' concurrent execution on Tue Nov 27 21:18:47 2018 with 2.191 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/3-overlapErrorAdjustment\n",
      "    ./oea.sh 1 > ./oea.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:49 2018 (2 seconds) with 2.191 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found 1 overlap error adjustment output files.\n",
      "-- Finished stage 'overlapErrorAdjustmentCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:49 2018 with 2.191 GB free disk space\n",
      "\n",
      "    cd unitigging/3-overlapErrorAdjustment\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/loadErates \\\n",
      "      -S ../../sample.seqStore \\\n",
      "      -O ../sample.ovlStore \\\n",
      "      -L ./oea.files \\\n",
      "    > ./oea.apply.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:49 2018 (in the blink of an eye) with 2.191 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- No report available.\n",
      "-- Finished stage 'updateOverlapStore', reset canuIteration.\n",
      "-- Finished stage 'unitig', reset canuIteration.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'bat' concurrent execution on Tue Nov 27 21:18:49 2018 with 2.191 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/4-unitigger\n",
      "    ./unitigger.sh 1 > ./unitigger.000001.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:49 2018 (furiously fast) with 2.19 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Unitigger finished successfully.\n",
      "-- Found, in version 1, after unitig construction:\n",
      "--   contigs:      47 sequences, total length 353360 bp (including 0 repeats of total length 0 bp).\n",
      "--   bubbles:      0 sequences, total length 0 bp.\n",
      "--   unassembled:  973 sequences, total length 1723250 bp.\n",
      "--\n",
      "-- Contig sizes based on genome size 2.1mbp:\n",
      "--\n",
      "--            NG (bp)  LG (contigs)    sum (bp)\n",
      "--         ----------  ------------  ----------\n",
      "--     10       12546             9      216891\n",
      "--\n",
      "-- Finished stage 'unitigCheck', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:49 2018 with 2.19 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/sqStoreCreatePartition \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -T  ./sample.ctgStore 1 \\\n",
      "      -b 15000 \\\n",
      "      -p 8 \\\n",
      "    > ./sample.ctgStore/partitionedReads.log 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:49 2018 (like a bat out of hell) with 2.185 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:49 2018 with 2.185 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/sqStoreCreatePartition \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -T  ./sample.utgStore 1 \\\n",
      "      -b 15000 \\\n",
      "      -p 8 \\\n",
      "    > ./sample.utgStore/partitionedReads.log 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:49 2018 (furiously fast) with 2.183 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Using slow alignment for consensus (iteration '0').\n",
      "-- Configured 1 contig and 1 unitig consensus jobs.\n",
      "-- Finished stage 'consensusConfigure', reset canuIteration.\n",
      "--\n",
      "--                            (tag)Concurrency\n",
      "--                     (tag)Threads          |\n",
      "--            (tag)Memory         |          |\n",
      "--        (tag)         |         |          |     total usage     algorithm\n",
      "--        -------  ------  --------   --------  -----------------  -----------------------------\n",
      "-- Local: cns        1 GB    3 CPUs x   1 job      1 GB    3 CPUs  (consensus)\n",
      "--\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'cns' concurrent execution on Tue Nov 27 21:18:50 2018 with 2.183 GB free disk space (2 processes; 1 concurrently)\n",
      "\n",
      "    cd unitigging/5-consensus\n",
      "    ./consensus.sh 1 > ./consensus.000001.out 2>&1\n",
      "    ./consensus.sh 2 > ./consensus.000002.out 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:57 2018 (7 seconds) with 2.179 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Using slow alignment for consensus (iteration '1').\n",
      "-- Configured 1 contig and 1 unitig consensus jobs.\n",
      "-- All 2 consensus jobs finished successfully.\n",
      "-- Finished stage 'consensusCheck', reset canuIteration.\n",
      "-- Using slow alignment for consensus (iteration '0').\n",
      "-- Configured 1 contig and 1 unitig consensus jobs.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:57 2018 with 2.179 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreLoad \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -T  ./sample.ctgStore 2 \\\n",
      "      -L ./5-consensus/ctgcns.files \\\n",
      "    > ./5-consensus/ctgcns.files.ctgStoreLoad.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:57 2018 (furiously fast) with 2.175 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:57 2018 with 2.175 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreLoad \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -T  ./sample.utgStore 2 \\\n",
      "      -L ./5-consensus/utgcns.files \\\n",
      "    > ./5-consensus/utgcns.files.utgStoreLoad.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:57 2018 (in the blink of an eye) with 2.174 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Purging consensus output after loading to ctgStore and/or utgStore.\n",
      "-- Purged 2 .cns outputs.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:57 2018 with 2.185 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreDump \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -T ./sample.ctgStore 2 \\\n",
      "      -sizes -s 2100000 \\\n",
      "    > ./sample.ctgStore/seqDB.v002.sizes.txt\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:57 2018 (fast as lightning) with 2.185 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Found, in version 2, after consensus generation:\n",
      "--   contigs:      47 sequences, total length 351969 bp (including 0 repeats of total length 0 bp).\n",
      "--   bubbles:      0 sequences, total length 0 bp.\n",
      "--   unassembled:  973 sequences, total length 1722848 bp.\n",
      "--\n",
      "-- Contig sizes based on genome size 2.1mbp:\n",
      "--\n",
      "--            NG (bp)  LG (contigs)    sum (bp)\n",
      "--         ----------  ------------  ----------\n",
      "--     10       12478             9      216453\n",
      "--\n",
      "-- Finished stage 'consensusLoad', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:57 2018 with 2.185 GB free disk space\n",
      "\n",
      "    cd unitigging\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreCoverageStat \\\n",
      "      -S ../sample.seqStore \\\n",
      "      -T  ./sample.ctgStore 2 \\\n",
      "      -s 2100000 \\\n",
      "      -o ./sample.ctgStore.coverageStat \\\n",
      "    > ./sample.ctgStore.coverageStat.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:57 2018 (like a bat out of hell) with 2.185 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Finished stage 'consensusAnalyze', reset canuIteration.\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:57 2018 with 2.185 GB free disk space\n",
      "\n",
      "    cd unitigging/4-unitigger\n",
      "    ./alignGFA.sh \\\n",
      "    > alignGFA.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:58 2018 (one second) with 2.185 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:58 2018 with 2.185 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreDump \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -T ./unitigging/sample.ctgStore 2 \\\n",
      "      -o ./sample.contigs \\\n",
      "      -layout \\\n",
      "    > ./sample.contigs.layout.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:58 2018 (furiously fast) with 2.181 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:58 2018 with 2.181 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreDump \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -T ./unitigging/sample.utgStore 2 \\\n",
      "      -o ./sample.unitigs \\\n",
      "      -layout \\\n",
      "    > ./sample.unitigs.layout.err 2>&1\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:58 2018 (fast as lightning) with 2.18 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:58 2018 with 2.18 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreDump \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -T ./unitigging/sample.ctgStore 2 \\\n",
      "      -consensus -fasta \\\n",
      "      -unassembled \\\n",
      "    > ./sample.unassembled.fasta\n",
      "    2> ./sample.unassembled.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:58 2018 (in the blink of an eye) with 2.178 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:58 2018 with 2.178 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreDump \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -T ./unitigging/sample.ctgStore 2 \\\n",
      "      -consensus -fasta \\\n",
      "      -contigs \\\n",
      "    > ./sample.contigs.fasta\n",
      "    2> ./sample.contigs.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:58 2018 (furiously fast) with 2.178 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "----------------------------------------\n",
      "-- Starting command on Tue Nov 27 21:18:58 2018 with 2.178 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/tgStoreDump \\\n",
      "      -S ./sample.seqStore \\\n",
      "      -T ./unitigging/sample.utgStore 2 \\\n",
      "      -consensus -fasta \\\n",
      "      -contigs \\\n",
      "    > ./sample.unitigs.fasta\n",
      "    2> ./sample.unitigs.err\n",
      "\n",
      "-- Finished on Tue Nov 27 21:18:58 2018 (like a bat out of hell) with 2.178 GB free disk space  !!! WARNING !!!\n",
      "----------------------------------------\n",
      "-- Finished stage 'generateOutputs', reset canuIteration.\n",
      "--\n",
      "-- Assembly 'sample' finished in '/home/jovyan/notebooks/data/sample/canu_output'.\n",
      "--\n",
      "-- Summary saved in 'sample.report'.\n",
      "--\n",
      "-- Sequences saved:\n",
      "--   Contigs       -> 'sample.contigs.fasta'\n",
      "--   Unassembled   -> 'sample.unassembled.fasta'\n",
      "--   Unitigs       -> 'sample.unitigs.fasta'\n",
      "--\n",
      "-- Read layouts saved:\n",
      "--   Contigs       -> 'sample.contigs.layout'.\n",
      "--   Unitigs       -> 'sample.unitigs.layout'.\n",
      "--\n",
      "-- Graphs saved:\n",
      "--   Contigs       -> 'sample.contigs.gfa'.\n",
      "--   Unitigs       -> 'sample.unitigs.gfa'.\n",
      "--\n",
      "-- Bye.\n"
     ]
    }
   ],
   "source": [
    "canu -p sample \\\n",
    "     -d data/sample/canu_output \\\n",
    "     genomeSize=2.1m \\\n",
    "     useGrid=false \\\n",
    "     minReadLength=50 \\\n",
    "     minOverlapLength=50 \\\n",
    "     corMemory=2 \\\n",
    "     corThreads=2 \\\n",
    "     maxMemory=6 \\\n",
    "     stopOnLowCoverage=1 \\\n",
    "     -nanopore-raw data/sample/reads.fastq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Racon\n",
    "\n",
    "[Racon](https://github.com/isovic/racon) is a consensus module to correct raw contigs generated by rapid assembly methods that do not include a consensus step. Canu results in FASTA format and raw reads in FASTQ are used in this step to generate a new FASTA contig file.\n",
    "\n",
    "Before running Racon, one must align the Canu output to the raw reads file and take the overlaps file in PAF as a parameter for Racon command. This can be done using for example minimap, a superfast aligner for ONT reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.010*0.80] collected minimizers\n",
      "[M::mm_idx_gen::0.017*1.37] sorted minimizers\n",
      "[M::main::0.018*1.37] loaded/built the index for 47 target sequence(s)\n",
      "[M::mm_mapopt_update::0.018*1.33] mid_occ = 5\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 47\n",
      "[M::mm_idx_stat::0.020*1.23] distinct minimizers: 61757 (93.56% are singletons); average occurrences: 1.067; average spacing: 5.340\n",
      "[M::worker_pipeline::0.606*2.32] mapped 3720 sequences\n",
      "[M::main] Version: 2.14-r892-dirty\n",
      "[M::main] CMD: minimap2 data/sample/canu_output/sample.contigs.fasta data/sample/reads.fastq\n",
      "[M::main] Real time: 0.610 sec; CPU: 1.408 sec; Peak RSS: 0.061 GB\n"
     ]
    }
   ],
   "source": [
    "minimap2 data/sample/canu_output/sample.contigs.fasta \\\n",
    "        data/sample/reads.fastq \\\n",
    "        > data/sample/aligned_reads.paf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The basic Racon parameters are the following:\n",
    "- **-t** : Number of threads\n",
    "- Raw reads (FASTQ)\n",
    "- Overlaps in (PAF)\n",
    "- Canu output (FASTA)\n",
    "- Racon output file name (FASTA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[racon::Polisher::initialize] loaded target sequences\n",
      "[racon::Polisher::initialize] loaded sequences\n",
      "[racon::Polisher::initialize] loaded overlaps\n",
      "[racon::Polisher::initialize] aligned overlap 785/785\n",
      "[racon::Polisher::initialize] transformed data into windows\n",
      "[racon::Polisher::polish] generated consensus for window 726/726\n"
     ]
    }
   ],
   "source": [
    "mkdir -p data/sample/racon_output\n",
    "racon -t 48 \\\n",
    "     data/sample/reads.fastq \\\n",
    "     data/sample/aligned_reads.paf \\\n",
    "     data/sample/canu_output/sample.contigs.fasta > data/sample/racon_output/sample_racon.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pilon (Requires Illumina reads)\n",
    "\n",
    "[Pilon](https:github.com/broadinstitute/pilon) is a tool that can be used to improve a draft assembly and find variation among species or strains. Pilon maps the reads from Illumina read to an assembled sequence and corrects the errors of the base, and the small insertions and deletions (indels). It requires as input a FASTA file and a BAM file of reads aligned to the input FASTA file. At this point, it will need the Racon contigs file and the reads file (ONT and Illumina reads). The Racon contigs file and the BAM file produced by de alignment of Illumina reads against that contigs file generate the final result of Pilon.\n",
    "\n",
    "In first place, Illumina reads have to be aligned against the Racon contigs file. BWA is used to index the reference file (Racon contigs) and BWA-MEM is used to perform the alignment of the Illumina reads. In order to have the BAM file required by Pilon, SAMtools is used to convert the BWA output in FASTA to the BAM format.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.00 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.07 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.00 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.03 sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: bwa index data/sample/racon_output/sample_racon.contigs.fasta\n",
      "[main] Real time: 0.154 sec; CPU: 0.104 sec\n"
     ]
    }
   ],
   "source": [
    "bwa index data/sample/racon_output/sample_racon.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 81792 sequences (20000386 bp)...\n",
      "[M::process] read 81780 sequences (20000192 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 5952, 6, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 421, 572)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1100)\n",
      "[M::mem_pestat] mean and std.dev: (451.48, 196.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1364)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81792 reads in 9.988 CPU sec, 5.098 real sec\n",
      "[M::process] read 81768 sequences (20000120 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5888, 7, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 417, 576)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1114)\n",
      "[M::mem_pestat] mean and std.dev: (451.15, 198.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1383)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81780 reads in 10.016 CPU sec, 5.349 real sec\n",
      "[M::process] read 81746 sequences (20000057 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5830, 9, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (305, 415, 569)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1097)\n",
      "[M::mem_pestat] mean and std.dev: (447.95, 200.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1361)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81768 reads in 10.176 CPU sec, 5.299 real sec\n",
      "[M::process] read 81760 sequences (20000155 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 5669, 6, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 422, 578)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1110)\n",
      "[M::mem_pestat] mean and std.dev: (455.17, 198.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1376)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81746 reads in 10.280 CPU sec, 5.248 real sec\n",
      "[M::process] read 81802 sequences (20000060 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5778, 8, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 421, 573)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1099)\n",
      "[M::mem_pestat] mean and std.dev: (451.71, 196.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1362)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81760 reads in 10.176 CPU sec, 5.255 real sec\n",
      "[M::process] read 81756 sequences (20000449 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5899, 6, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (309, 425, 576)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1110)\n",
      "[M::mem_pestat] mean and std.dev: (454.61, 197.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1377)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81802 reads in 10.208 CPU sec, 5.346 real sec\n",
      "[M::process] read 81782 sequences (20000374 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5779, 4, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (309, 418, 573)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1101)\n",
      "[M::mem_pestat] mean and std.dev: (451.44, 196.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1365)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81756 reads in 9.692 CPU sec, 5.083 real sec\n",
      "[M::process] read 81774 sequences (20000283 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5833, 15, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 424, 580)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1120)\n",
      "[M::mem_pestat] mean and std.dev: (456.15, 200.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1390)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (87, 521, 4949)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 14673)\n",
      "[M::mem_pestat] mean and std.dev: (1526.73, 2126.87)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 19535)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81782 reads in 9.884 CPU sec, 5.127 real sec\n",
      "[M::process] read 81814 sequences (20000462 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5892, 2, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 426, 578)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1110)\n",
      "[M::mem_pestat] mean and std.dev: (455.16, 196.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1376)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81774 reads in 10.192 CPU sec, 5.244 real sec\n",
      "[M::process] read 81862 sequences (20000162 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5792, 4, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 416, 568)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1092)\n",
      "[M::mem_pestat] mean and std.dev: (446.55, 193.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1354)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81814 reads in 9.952 CPU sec, 5.180 real sec\n",
      "[M::process] read 81812 sequences (20000381 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5853, 9, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 414, 566)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1086)\n",
      "[M::mem_pestat] mean and std.dev: (445.34, 192.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1346)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81862 reads in 9.824 CPU sec, 4.879 real sec\n",
      "[M::process] read 81770 sequences (20000499 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5882, 10, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (305, 416, 569)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1097)\n",
      "[M::mem_pestat] mean and std.dev: (446.31, 193.52)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1361)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (141, 517, 521)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1281)\n",
      "[M::mem_pestat] mean and std.dev: (308.75, 196.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1661)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81812 reads in 9.644 CPU sec, 4.809 real sec\n",
      "[M::process] read 81812 sequences (20000109 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5873, 3, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 416, 568)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1088)\n",
      "[M::mem_pestat] mean and std.dev: (446.22, 192.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1348)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81770 reads in 10.900 CPU sec, 6.026 real sec\n",
      "[M::process] read 81848 sequences (20000367 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5825, 5, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 418, 568)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1088)\n",
      "[M::mem_pestat] mean and std.dev: (449.35, 195.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1348)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81812 reads in 10.084 CPU sec, 5.216 real sec\n",
      "[M::process] read 81806 sequences (20000071 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5993, 6, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 418, 558)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1060)\n",
      "[M::mem_pestat] mean and std.dev: (444.15, 189.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1311)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81848 reads in 9.980 CPU sec, 4.950 real sec\n",
      "[M::process] read 81796 sequences (20000486 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5893, 4, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (313, 422, 566)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1072)\n",
      "[M::mem_pestat] mean and std.dev: (448.94, 187.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1325)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81806 reads in 9.844 CPU sec, 4.995 real sec\n",
      "[M::process] read 81852 sequences (20000361 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 5837, 5, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 412, 558)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1058)\n",
      "[M::mem_pestat] mean and std.dev: (442.81, 188.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1308)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81796 reads in 9.812 CPU sec, 4.892 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5842, 6, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (305, 413, 561)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1073)\n",
      "[M::mem_pestat] mean and std.dev: (442.19, 189.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1329)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81852 reads in 10.880 CPU sec, 5.992 real sec\n",
      "[M::process] read 81676 sequences (20000470 bp)...\n",
      "[M::process] read 81738 sequences (20000005 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5782, 4, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 428, 588)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1136)\n",
      "[M::mem_pestat] mean and std.dev: (462.29, 203.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1410)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81676 reads in 9.828 CPU sec, 4.926 real sec\n",
      "[M::process] read 81680 sequences (20000081 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5905, 7, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 424, 587)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1137)\n",
      "[M::mem_pestat] mean and std.dev: (461.02, 205.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1412)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81738 reads in 9.800 CPU sec, 4.879 real sec\n",
      "[M::process] read 81676 sequences (20000475 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5764, 8, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 425, 592)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1152)\n",
      "[M::mem_pestat] mean and std.dev: (463.93, 209.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1432)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81680 reads in 9.960 CPU sec, 5.014 real sec\n",
      "[M::process] read 81682 sequences (20000400 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5840, 11, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 423, 582)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1126)\n",
      "[M::mem_pestat] mean and std.dev: (457.40, 203.07)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1398)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 427, 521)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1087)\n",
      "[M::mem_pestat] mean and std.dev: (325.67, 141.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1370)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81676 reads in 9.880 CPU sec, 4.982 real sec\n",
      "[M::process] read 81746 sequences (20000199 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5889, 9, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 428, 587)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1133)\n",
      "[M::mem_pestat] mean and std.dev: (461.99, 207.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1406)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81682 reads in 9.652 CPU sec, 4.853 real sec\n",
      "[M::process] read 81714 sequences (20000155 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5815, 10, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (315, 426, 588)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1134)\n",
      "[M::mem_pestat] mean and std.dev: (462.49, 206.44)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1407)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (86, 406, 4654)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13790)\n",
      "[M::mem_pestat] mean and std.dev: (1649.40, 2217.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18358)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81746 reads in 9.604 CPU sec, 4.826 real sec\n",
      "[M::process] read 81740 sequences (20000239 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5862, 9, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 427, 592)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1156)\n",
      "[M::mem_pestat] mean and std.dev: (463.32, 208.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1438)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81714 reads in 9.932 CPU sec, 4.954 real sec\n",
      "[M::process] read 81670 sequences (20000022 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 5907, 8, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 429, 581)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1115)\n",
      "[M::mem_pestat] mean and std.dev: (460.03, 203.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1382)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81740 reads in 10.576 CPU sec, 5.368 real sec\n",
      "[M::process] read 81694 sequences (20000347 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5876, 7, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 427, 587)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1133)\n",
      "[M::mem_pestat] mean and std.dev: (463.02, 207.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1406)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81670 reads in 10.016 CPU sec, 4.982 real sec\n",
      "[M::process] read 81634 sequences (20000486 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5731, 10, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (317, 428, 588)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1130)\n",
      "[M::mem_pestat] mean and std.dev: (462.24, 204.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1401)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (217, 521, 4954)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 14428)\n",
      "[M::mem_pestat] mean and std.dev: (2164.80, 2290.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 19165)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81694 reads in 9.884 CPU sec, 4.917 real sec\n",
      "[M::process] read 81744 sequences (20000445 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 5938, 9, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (317, 429, 586)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1124)\n",
      "[M::mem_pestat] mean and std.dev: (462.68, 201.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1393)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81634 reads in 10.428 CPU sec, 5.204 real sec\n",
      "[M::process] read 81738 sequences (20000031 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 5869, 11, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (309, 416, 572)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1098)\n",
      "[M::mem_pestat] mean and std.dev: (452.07, 200.24)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1361)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (80, 510, 4598)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 13634)\n",
      "[M::mem_pestat] mean and std.dev: (1486.82, 2040.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 18152)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81744 reads in 9.668 CPU sec, 4.880 real sec\n",
      "[M::process] read 81648 sequences (20000179 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5907, 6, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (315, 421, 577)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1101)\n",
      "[M::mem_pestat] mean and std.dev: (457.64, 199.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1363)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81738 reads in 9.832 CPU sec, 4.900 real sec\n",
      "[M::process] read 81720 sequences (20000440 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5876, 6, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 426, 574)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1098)\n",
      "[M::mem_pestat] mean and std.dev: (455.59, 198.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1360)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81648 reads in 9.844 CPU sec, 4.926 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 5945, 6, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (313, 425, 576)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1102)\n",
      "[M::mem_pestat] mean and std.dev: (455.22, 197.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1365)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81720 reads in 10.752 CPU sec, 5.737 real sec\n",
      "[M::process] read 81738 sequences (20000138 bp)...\n",
      "[M::process] read 81734 sequences (20000396 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5828, 6, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 428, 582)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1122)\n",
      "[M::mem_pestat] mean and std.dev: (458.13, 201.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1392)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81738 reads in 10.276 CPU sec, 5.438 real sec\n",
      "[M::process] read 81680 sequences (20000320 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 5926, 9, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 421, 581)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1129)\n",
      "[M::mem_pestat] mean and std.dev: (456.63, 203.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1403)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 81734 reads in 10.092 CPU sec, 5.125 real sec\n",
      "[M::process] read 32372 sequences (7928841 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 5775, 10, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (309, 424, 580)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1122)\n",
      "[M::mem_pestat] mean and std.dev: (455.87, 201.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1393)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (87, 179, 521)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1389)\n",
      "[M::mem_pestat] mean and std.dev: (207.50, 185.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1823)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 81680 reads in 10.428 CPU sec, 5.421 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 2357, 1, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 427, 578)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1114)\n",
      "[M::mem_pestat] mean and std.dev: (458.20, 203.38)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1382)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 32372 reads in 3.848 CPU sec, 2.051 real sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: bwa mem -t 2 data/sample/racon_output/sample_racon.contigs.fasta data/sample/short/reads_1.fastq.gz data/sample/short/reads_2.fastq.gz\n",
      "[main] Real time: 197.020 sec; CPU: 366.888 sec\n",
      "[bam_sort_core] merging from 3 files...\n"
     ]
    }
   ],
   "source": [
    "mkdir -p data/sample/bwa_output\n",
    "#Data not included in repository\n",
    "bwa mem -t 2 \\\n",
    "        data/sample/racon_output/sample_racon.contigs.fasta  \\\n",
    "        data/sample/short/reads_1.fastq.gz \\\n",
    "        data/sample/short/reads_2.fastq.gz \\\n",
    "        | samtools view -S -b -u - | samtools sort -o data/sample/bwa_output/bwa_aligned_reads.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before running Pilon, the BWA alignment is indexed using SAMtools:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "samtools index data/sample/bwa_output/bwa_aligned_reads.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pilon is run as a Java .jar executable. Some options can be added to improve the performance of Pilon before specifying the .jar file. Whichever the case, Pilon works better in terms of execution time when it is run on an environment with more available RAM and threads. The parameters used in this Pilon run are the following:\n",
    "\n",
    "- **--threads**: Number of threads\n",
    "- **--genome**: The FASTA input file (Racon contigs)\n",
    "- **--bam**: BAM file (generated by BWA and SAMtools)\n",
    "- **--outdir** and **--output**: Output directory and filename "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pilon version 1.22 Wed Mar 15 16:38:30 2017 -0400\n",
      "Genome: data/sample/racon_output/sample_racon.contigs.fasta\n",
      "Fixing snps, indels, gaps, local\n",
      "Input genome size: 356234\n",
      "Scanning BAMs\n",
      "data/sample/bwa_output/bwa_aligned_reads.bam: 2976579 reads, 0 filtered, 463011 mapped, 444443 proper, 1788 stray, FR 100% 454+/-209, max 1081 frags\n",
      "Processing tig00000242:1-13001\n",
      "Processing tig00000026:1-11631\n",
      "tig00000242:1-13001 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 303\n",
      "Total Reads: 18587, Coverage: 303, minDepth: 30\n",
      "Confirmed 12588 of 13001 bases (96.82%)\n",
      "Corrected 84 snps; 0 ambiguous bases; corrected 61 small insertions totaling 83 bases, 152 small deletions totaling 254 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000242:540-975 0 -0 +0 NoSolution\n",
      "fix break: tig00000242:1687-1688 1293 -781 +767 BreakFix\n",
      "# fix break: tig00000242:2001 0 -0 +0 NoSolution\n",
      "fix break: tig00000242:3676-3826 3279 -800 +786 BreakFix\n",
      "fix break: tig00000242:5040-5041 4691 -670 +658 BreakFix\n",
      "# fix break: tig00000242:5259-5360 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:6043 0 -0 +0 NoSolution\n",
      "fix break: tig00000242:6324-6644 6047 -1005 +987 BreakFix\n",
      "fix break: tig00000242:7295-7725 6934 -1149 +1116 BreakFix\n",
      "# fix break: tig00000242:8031 0 -0 +0 NoSolution\n",
      "fix break: tig00000242:8436-8639 8031 -1000 +995 BreakFix\n",
      "# fix break: tig00000242:8856-8964 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:9612-9714 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:10188 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:10582-10613 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:11015-11016 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:11544-11545 0 -0 +0 NoSolution\n",
      "# fix break: tig00000242:12096-12236 0 -0 +0 NoSolution\n",
      "Finished processing tig00000242:1-13001\n",
      "Processing tig00000091:1-786\n",
      "tig00000091:1-786 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 295\n",
      "Total Reads: 2242, Coverage: 295, minDepth: 30\n",
      "Confirmed 753 of 786 bases (95.80%)\n",
      "Corrected 4 snps; 0 ambiguous bases; corrected 3 small insertions totaling 3 bases, 12 small deletions totaling 26 bases\n",
      "Finished processing tig00000091:1-786\n",
      "Processing tig00000245:1-1208\n",
      "tig00000245:1-1208 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 241\n",
      "Total Reads: 2056, Coverage: 241, minDepth: 24\n",
      "Confirmed 1162 of 1208 bases (96.19%)\n",
      "Corrected 5 snps; 0 ambiguous bases; corrected 5 small insertions totaling 6 bases, 21 small deletions totaling 36 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000245:335-337 0 -0 +0 NoSolution\n",
      "# fix break: tig00000245:770-816 0 -0 +0 NoSolution\n",
      "Finished processing tig00000245:1-1208\n",
      "Processing tig00000250:1-1078\n",
      "tig00000250:1-1078 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 208\n",
      "Total Reads: 1648, Coverage: 208, minDepth: 21\n",
      "Confirmed 1010 of 1078 bases (93.69%)\n",
      "Corrected 13 snps; 0 ambiguous bases; corrected 6 small insertions totaling 7 bases, 33 small deletions totaling 49 bases\n",
      "Finished processing tig00000250:1-1078\n",
      "Processing tig00000095:1-2644\n",
      "tig00000026:1-11631 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 272\n",
      "Total Reads: 15705, Coverage: 272, minDepth: 27\n",
      "Confirmed 11162 of 11631 bases (95.97%)\n",
      "Corrected 84 snps; 0 ambiguous bases; corrected 53 small insertions totaling 69 bases, 204 small deletions totaling 321 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000026:933-954 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:1380-1777 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:2265-2940 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:4424 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:4739 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:5181-5356 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:5850-6112 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:6384-6559 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:6946-7062 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:7357-7449 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:7913-8112 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:8420 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:9203 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:9638 0 -0 +0 NoSolution\n",
      "# fix break: tig00000026:9851-9939 0 -0 +0 NoSolution\n",
      "fix break: tig00000026:10330 9938 -689 +671 BreakFix\n",
      "# fix break: tig00000026:10556-10597 0 -0 +0 NoSolution\n",
      "Finished processing tig00000026:1-11631\n",
      "Processing tig00000116:1-7864\n",
      "tig00000095:1-2644 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 199\n",
      "Total Reads: 2733, Coverage: 199, minDepth: 20\n",
      "Confirmed 2508 of 2644 bases (94.86%)\n",
      "Corrected 17 snps; 0 ambiguous bases; corrected 29 small insertions totaling 37 bases, 41 small deletions totaling 85 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000095:364-1130 0 -0 +0 NoSolution\n",
      "# fix break: tig00000095:1454-1455 0 -0 +0 NoSolution\n",
      "# fix break: tig00000095:1904-2244 0 -0 +0 NoSolution\n",
      "Finished processing tig00000095:1-2644\n",
      "Processing tig00000193:1-2818\n",
      "tig00000193:1-2818 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 253\n",
      "Total Reads: 3763, Coverage: 253, minDepth: 25\n",
      "Confirmed 2663 of 2818 bases (94.50%)\n",
      "Corrected 30 snps; 0 ambiguous bases; corrected 24 small insertions totaling 35 bases, 54 small deletions totaling 101 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000193:127-329 0 -0 +0 NoSolution\n",
      "# fix break: tig00000193:771-1218 0 -0 +0 NoSolution\n",
      "# fix break: tig00000193:1463-1464 0 -0 +0 NoSolution\n",
      "# fix break: tig00000193:1857-2101 0 -0 +0 NoSolution\n",
      "# fix break: tig00000193:2369-2370 0 -0 +0 NoSolution\n",
      "Finished processing tig00000193:1-2818\n",
      "Processing tig00000033:1-35334\n",
      "tig00000116:1-7864 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 336\n",
      "Total Reads: 13555, Coverage: 336, minDepth: 34\n",
      "Confirmed 7580 of 7864 bases (96.39%)\n",
      "Corrected 70 snps; 0 ambiguous bases; corrected 45 small insertions totaling 60 bases, 98 small deletions totaling 153 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000116:1742-1790 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:2758-2882 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:3209 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:3890-4681 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:5080 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:5376-5528 0 -0 +0 NoSolution\n",
      "fix break: tig00000116:5924-6156 5528 -1005 +989 BreakFix\n",
      "# fix break: tig00000116:6420-6532 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:6783 0 -0 +0 NoSolution\n",
      "# fix break: tig00000116:7091-7418 0 -0 +0 NoSolution\n",
      "Finished processing tig00000116:1-7864\n",
      "Processing tig00000114:1-2710\n",
      "tig00000114:1-2710 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 351\n",
      "Total Reads: 5113, Coverage: 351, minDepth: 35\n",
      "Confirmed 2599 of 2710 bases (95.90%)\n",
      "Corrected 23 snps; 0 ambiguous bases; corrected 25 small insertions totaling 30 bases, 33 small deletions totaling 61 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "fix break: tig00000114:1340-1535 998 -934 +928 BreakFix\n",
      "# fix break: tig00000114:1776-2403 0 -0 +0 NoSolution\n",
      "Finished processing tig00000114:1-2710\n",
      "Processing tig00000049:1-1443\n",
      "tig00000049:1-1443 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 220\n",
      "Total Reads: 2125, Coverage: 220, minDepth: 22\n",
      "Confirmed 1360 of 1443 bases (94.25%)\n",
      "Corrected 17 snps; 0 ambiguous bases; corrected 10 small insertions totaling 19 bases, 33 small deletions totaling 52 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000049:364 0 -0 +0 NoSolution\n",
      "# fix break: tig00000049:567-742 0 -0 +0 NoSolution\n",
      "Finished processing tig00000049:1-1443\n",
      "Processing tig00000244:1-6926\n",
      "tig00000244:1-6926 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 330\n",
      "Total Reads: 11326, Coverage: 330, minDepth: 33\n",
      "Confirmed 6676 of 6926 bases (96.39%)\n",
      "Corrected 62 snps; 0 ambiguous bases; corrected 36 small insertions totaling 45 bases, 93 small deletions totaling 152 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000244:419-535 0 -0 +0 NoSolution\n",
      "fix break: tig00000244:1322-1441 993 -841 +830 BreakFix\n",
      "# fix break: tig00000244:1776 0 -0 +0 NoSolution\n",
      "# fix break: tig00000244:2273-2467 0 -0 +0 NoSolution\n",
      "fix break: tig00000244:2679-2680 2277 -798 +781 BreakFix\n",
      "# fix break: tig00000244:3222-3261 0 -0 +0 NoSolution\n",
      "fix break: tig00000244:3984-4290 3704 -904 +893 BreakFix\n",
      "# fix break: tig00000244:4749-5120 0 -0 +0 NoSolution\n",
      "# fix break: tig00000244:5338-5394 0 -0 +0 NoSolution\n",
      "# fix break: tig00000244:5742 0 -0 +0 NoSolution\n",
      "# fix break: tig00000244:6355-6644 0 -0 +0 NoSolution\n",
      "Finished processing tig00000244:1-6926\n",
      "Processing tig00000153:1-26377\n",
      "tig00000033:1-35334 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 239\n",
      "Total Reads: 39618, Coverage: 239, minDepth: 24\n",
      "Confirmed 34307 of 35334 bases (97.09%)\n",
      "Corrected 169 snps; 0 ambiguous bases; corrected 150 small insertions totaling 184 bases, 461 small deletions totaling 703 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000033:1005-1528 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:2902 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:3262-3376 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:3732-3836 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:4058-4059 3673 -760 +744 BreakFix\n",
      "# fix break: tig00000033:4317 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:4610-4663 4320 -626 +612 BreakFix\n",
      "# fix break: tig00000033:4944-4945 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:5965 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:6700-7170 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:7435-7462 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:7826-7827 7440 -732 +721 BreakFix\n",
      "fix break: tig00000033:8079 7828 -466 +458 BreakFix\n",
      "fix break: tig00000033:8293 7975 -706 +699 BreakFix\n",
      "# fix break: tig00000033:9152-9534 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:10174-10410 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:10656-11349 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:11944 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:12249 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:12631-12632 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:13012 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:14087-14289 13758 -932 +911 BreakFix\n",
      "# fix break: tig00000033:14561-14689 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:14919 14516 -782 +778 BreakFix\n",
      "fix break: tig00000033:16076-16170 15719 -809 +794 BreakFix\n",
      "# fix break: tig00000033:17042-17466 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:17809-17857 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:18483-18705 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:19143 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:19421-19468 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:21638-21884 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:22489-23094 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:23329 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:23664-23827 23333 -846 +825 BreakFix\n",
      "# fix break: tig00000033:24091-24158 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:24795 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:25879 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:26224-26226 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:26837-27978 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:28326-28739 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:28988 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:29200-29552 0 -0 +0 NoSolution\n",
      "fix break: tig00000033:30069-30437 29757 -1059 +1046 BreakFix\n",
      "# fix break: tig00000033:31095 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:32161-32438 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:32723-33176 0 -0 +0 NoSolution\n",
      "# fix break: tig00000033:33549-34207 0 -0 +0 NoSolution\n",
      "Finished processing tig00000033:1-35334\n",
      "Processing tig00000072:1-2934\n",
      "tig00000072:1-2934 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 207\n",
      "Total Reads: 3479, Coverage: 207, minDepth: 21\n",
      "Confirmed 2845 of 2934 bases (96.97%)\n",
      "Corrected 14 snps; 0 ambiguous bases; corrected 20 small insertions totaling 21 bases, 30 small deletions totaling 55 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000072:183-184 0 -0 +0 NoSolution\n",
      "# fix break: tig00000072:459-461 0 -0 +0 NoSolution\n",
      "fix break: tig00000072:1546 1171 -738 +727 BreakFix\n",
      "# fix break: tig00000072:1844-1908 0 -0 +0 NoSolution\n",
      "Finished processing tig00000072:1-2934\n",
      "Processing tig00000092:1-604\n",
      "tig00000092:1-604 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 172\n",
      "Total Reads: 1236, Coverage: 172, minDepth: 17\n",
      "Confirmed 573 of 604 bases (94.87%)\n",
      "Corrected 4 snps; 0 ambiguous bases; corrected 5 small insertions totaling 5 bases, 17 small deletions totaling 22 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000092:197 0 -0 +0 NoSolution\n",
      "Finished processing tig00000092:1-604\n",
      "Processing tig00000203:1-702\n",
      "tig00000203:1-702 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 175\n",
      "Total Reads: 1165, Coverage: 175, minDepth: 18\n",
      "Confirmed 661 of 702 bases (94.16%)\n",
      "Corrected 11 snps; 0 ambiguous bases; corrected 7 small insertions totaling 12 bases, 13 small deletions totaling 23 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000203:292-581 0 -0 +0 NoSolution\n",
      "Finished processing tig00000203:1-702\n",
      "Processing tig00001190:1-12445\n",
      "tig00001190:1-12445 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 52\n",
      "Total Reads: 3669, Coverage: 52, minDepth: 5\n",
      "Confirmed 4200 of 12445 bases (33.75%)\n",
      "Corrected 67 snps; 0 ambiguous bases; corrected 34 small insertions totaling 51 bases, 75 small deletions totaling 113 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00001190:679-2461 0 -0 +0 NoSolution\n",
      "# fix break: tig00001190:2799-3133 0 -0 +0 NoSolution\n",
      "# fix break: tig00001190:3552-3563 0 -0 +0 NoSolution\n",
      "# fix break: tig00001190:3960-5998 0 -0 +0 NoSolution\n",
      "Finished processing tig00001190:1-12445\n",
      "Processing tig00000053:1-12634\n",
      "tig00000153:1-26377 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 361\n",
      "Total Reads: 45683, Coverage: 361, minDepth: 36\n",
      "Confirmed 25162 of 26377 bases (95.39%)\n",
      "Corrected 339 snps; 0 ambiguous bases; corrected 210 small insertions totaling 273 bases, 437 small deletions totaling 657 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000153:780-1115 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:1850-2131 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:2989-3145 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:3908 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:4152-4486 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:4802-5024 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:5547-5697 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:6305-6427 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:6662 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:7057-8134 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:8555-8788 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:10701-10777 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:11197-11496 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:11793-11837 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:12176-12517 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:12815-13067 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:13290-13535 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:13793-14362 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:14593-15198 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:15438 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:16126-16898 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:17108-17322 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:18525-18738 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:19704 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:19942-20647 0 -0 +0 NoSolution\n",
      "fix break: tig00000153:21049-21050 20650 -779 +775 BreakFix\n",
      "fix break: tig00000153:22221 21952 -635 +626 BreakFix\n",
      "# fix break: tig00000153:23496 0 -0 +0 NoSolution\n",
      "# fix break: tig00000153:23869 0 -0 +0 NoSolution\n",
      "fix break: tig00000153:24138-24307 23736 -916 +906 BreakFix\n",
      "fix break: tig00000153:24963-25178 24616 -954 +938 BreakFix\n",
      "# fix break: tig00000153:25513 0 -0 +0 NoSolution\n",
      "Finished processing tig00000153:1-26377\n",
      "Processing tig00000009:1-18553\n",
      "tig00000053:1-12634 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 372\n",
      "Total Reads: 22750, Coverage: 372, minDepth: 37\n",
      "Confirmed 12110 of 12634 bases (95.85%)\n",
      "Corrected 135 snps; 0 ambiguous bases; corrected 77 small insertions totaling 102 bases, 175 small deletions totaling 277 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000053:395-651 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:1026 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:1262-1410 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:1657-1845 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:2100-2699 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:2921-3086 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:3333-3650 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:3958-4383 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:4667-4882 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:5181 0 -0 +0 NoSolution\n",
      "fix break: tig00000053:5470-5512 5096 -791 +787 BreakFix\n",
      "# fix break: tig00000053:5870 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:6075-6076 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:6518-6777 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:7560 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:8560-8995 0 -0 +0 NoSolution\n",
      "# fix break: tig00000053:9220 0 -0 +0 NoSolution\n",
      "fix break: tig00000053:9598-9599 9221 -664 +658 BreakFix\n",
      "fix break: tig00000053:10130-10169 9810 -756 +748 BreakFix\n",
      "fix break: tig00000053:10691 10396 -694 +686 BreakFix\n",
      "# fix break: tig00000053:11289-11290 0 -0 +0 NoSolution\n",
      "fix break: tig00000053:11577-11859 11175 -1062 +1031 BreakFix\n",
      "# fix break: tig00000053:12236 0 -0 +0 NoSolution\n",
      "Finished processing tig00000053:1-12634\n",
      "Processing tig00000223:1-13369\n",
      "tig00000009:1-18553 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 304\n",
      "Total Reads: 26682, Coverage: 304, minDepth: 30\n",
      "Confirmed 17986 of 18553 bases (96.94%)\n",
      "Corrected 105 snps; 1 ambiguous bases; corrected 104 small insertions totaling 132 bases, 246 small deletions totaling 349 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000009:603-921 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:1259 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:1588-1782 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:2323 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:2527-2586 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:2843 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:3130-3131 0 -0 +0 NoSolution\n",
      "fix break: tig00000009:3602-3603 3198 -795 +780 BreakFix\n",
      "# fix break: tig00000009:3851-4065 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:4285-4708 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:4981-5283 0 -0 +0 NoSolution\n",
      "fix break: tig00000009:5887-5927 5480 -717 +703 BreakFix\n",
      "# fix break: tig00000009:6957-7045 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:7423 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:7654-7758 0 -0 +0 NoSolution\n",
      "fix break: tig00000009:9131 9052 -481 +477 BreakFix\n",
      "# fix break: tig00000009:10090-10160 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:10416 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:11195-11196 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:11736 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:12694-12847 0 -0 +0 NoSolution\n",
      "fix break: tig00000009:13613-13775 13238 -909 +900 BreakFix\n",
      "# fix break: tig00000009:14508-14537 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:15050-15259 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:15707-15971 0 -0 +0 NoSolution\n",
      "# fix break: tig00000009:17033 0 -0 +0 NoSolution\n",
      "fix break: tig00000009:17253-17336 16888 -832 +825 BreakFix\n",
      "# fix break: tig00000009:18037 0 -0 +0 NoSolution\n",
      "Finished processing tig00000009:1-18553\n",
      "Processing tig00000143:1-1261\n",
      "tig00000143:1-1261 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 255\n",
      "Total Reads: 2164, Coverage: 255, minDepth: 26\n",
      "Confirmed 1208 of 1261 bases (95.80%)\n",
      "Corrected 12 snps; 0 ambiguous bases; corrected 16 small insertions totaling 21 bases, 14 small deletions totaling 25 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000143:461-767 0 -0 +0 NoSolution\n",
      "Finished processing tig00000143:1-1261\n",
      "Processing tig00000119:1-2092\n",
      "tig00000119:1-2092 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 271\n",
      "Total Reads: 3224, Coverage: 271, minDepth: 27\n",
      "Confirmed 2019 of 2092 bases (96.51%)\n",
      "Corrected 10 snps; 0 ambiguous bases; corrected 11 small insertions totaling 12 bases, 32 small deletions totaling 51 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000119:517-890 0 -0 +0 NoSolution\n",
      "fix break: tig00000119:1428-1776 1042 -990 +974 BreakFix\n",
      "Finished processing tig00000119:1-2092\n",
      "Processing tig00000003:1-16669\n",
      "tig00000223:1-13369 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 274\n",
      "Total Reads: 18685, Coverage: 274, minDepth: 27\n",
      "Confirmed 12968 of 13369 bases (97.00%)\n",
      "Corrected 61 snps; 0 ambiguous bases; corrected 54 small insertions totaling 64 bases, 157 small deletions totaling 244 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000223:850-851 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:1527 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:1740-1742 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:2241-2243 0 -0 +0 NoSolution\n",
      "fix break: tig00000223:3299-3683 2943 -1076 +1058 BreakFix\n",
      "# fix break: tig00000223:3961-4002 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:4438-4468 0 -0 +0 NoSolution TandemRepeat 72\n",
      "# fix break: tig00000223:4718 0 -0 +0 NoSolution TandemRepeat 72\n",
      "# fix break: tig00000223:5859-7412 0 -0 +0 NoSolution TandemRepeat 72\n",
      "# fix break: tig00000223:7675-7749 0 -0 +0 NoSolution TandemRepeat 108\n",
      "# fix break: tig00000223:7979-7980 0 -0 +0 NoSolution TandemRepeat 108\n",
      "# fix break: tig00000223:8474 0 -0 +0 NoSolution TandemRepeat 108\n",
      "# fix break: tig00000223:8812-9114 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:10631-10853 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:11790-11791 0 -0 +0 NoSolution\n",
      "# fix break: tig00000223:12184-12427 0 -0 +0 NoSolution\n",
      "Finished processing tig00000223:1-13369\n",
      "Processing tig00000127:1-1448\n",
      "tig00000127:1-1448 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 184\n",
      "Total Reads: 1856, Coverage: 184, minDepth: 18\n",
      "Confirmed 1373 of 1448 bases (94.82%)\n",
      "Corrected 18 snps; 0 ambiguous bases; corrected 6 small insertions totaling 7 bases, 32 small deletions totaling 51 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000127:614 0 -0 +0 NoSolution\n",
      "# fix break: tig00000127:873-930 0 -0 +0 NoSolution\n",
      "Finished processing tig00000127:1-1448\n",
      "Processing tig00000123:1-19484\n",
      "tig00000123:1-19484 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 192\n",
      "Total Reads: 18396, Coverage: 192, minDepth: 19\n",
      "Confirmed 18852 of 19484 bases (96.76%)\n",
      "Corrected 114 snps; 0 ambiguous bases; corrected 86 small insertions totaling 104 bases, 278 small deletions totaling 427 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000123:3133-3161 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:3367-3737 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:4260-5044 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:5346-5592 0 -0 +0 NoSolution\n",
      "fix break: tig00000123:5827-5885 5456 -812 +787 BreakFix\n",
      "# fix break: tig00000123:6178-6568 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:7557 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:7825-8104 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:8410-8747 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:8986 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:9620-9782 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:10331 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:10901-10902 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:11253-11321 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:11956 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:12180-12269 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:12776-12778 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:13700-13818 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:14642-15043 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:15247 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:15780-15988 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:16429-16488 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:16933 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:17313 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:17985-18451 0 -0 +0 NoSolution\n",
      "# fix break: tig00000123:18888-18907 0 -0 +0 NoSolution\n",
      "Finished processing tig00000123:1-19484\n",
      "Processing tig00000088:1-796\n",
      "tig00000088:1-796 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 190\n",
      "Total Reads: 1202, Coverage: 190, minDepth: 19\n",
      "Confirmed 757 of 796 bases (95.10%)\n",
      "Corrected 2 snps; 0 ambiguous bases; corrected 10 small insertions totaling 12 bases, 14 small deletions totaling 27 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000088:304-466 0 -0 +0 NoSolution\n",
      "Finished processing tig00000088:1-796\n",
      "Processing tig00000008:1-11043\n",
      "tig00000008:1-11043 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 239\n",
      "Total Reads: 12518, Coverage: 239, minDepth: 24\n",
      "Confirmed 10271 of 11043 bases (93.01%)\n",
      "Corrected 63 snps; 0 ambiguous bases; corrected 46 small insertions totaling 72 bases, 183 small deletions totaling 335 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000008:698-909 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:1140-2415 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:2693-2783 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:3077 0 -0 +0 NoSolution\n",
      "fix break: tig00000008:3326-3448 3009 -811 +797 BreakFix\n",
      "# fix break: tig00000008:4077-4115 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:4538-4539 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:4893-4912 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:5328-5359 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:5615-5721 0 -0 +0 NoSolution\n",
      "fix break: tig00000008:6292 5962 -719 +708 BreakFix\n",
      "# fix break: tig00000008:6680-7206 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:7853-7868 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:8162-8576 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:9311-9554 0 -0 +0 NoSolution\n",
      "# fix break: tig00000008:9816-10183 0 -0 +0 NoSolution\n",
      "Finished processing tig00000008:1-11043\n",
      "Processing tig00000201:1-6022\n",
      "tig00000201:1-6022 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 192\n",
      "Total Reads: 6096, Coverage: 192, minDepth: 19\n",
      "Confirmed 5659 of 6022 bases (93.97%)\n",
      "Corrected 109 snps; 0 ambiguous bases; corrected 50 small insertions totaling 60 bases, 128 small deletions totaling 204 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000201:1397-1414 0 -0 +0 NoSolution\n",
      "# fix break: tig00000201:1616-1702 0 -0 +0 NoSolution\n",
      "# fix break: tig00000201:1970-2573 0 -0 +0 NoSolution\n",
      "# fix break: tig00000201:2952-3425 0 -0 +0 NoSolution\n",
      "# fix break: tig00000201:3725-4394 0 -0 +0 NoSolution\n",
      "Finished processing tig00000201:1-6022\n",
      "Processing tig00000182:1-2133\n",
      "tig00000182:1-2133 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 346\n",
      "Total Reads: 4287, Coverage: 346, minDepth: 35\n",
      "Confirmed 2005 of 2133 bases (94.00%)\n",
      "Corrected 28 snps; 0 ambiguous bases; corrected 15 small insertions totaling 24 bases, 50 small deletions totaling 85 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000182:1069-1466 0 -0 +0 NoSolution\n",
      "Finished processing tig00000182:1-2133\n",
      "Processing tig00000148:1-3888\n",
      "tig00000148:1-3888 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 426\n",
      "Total Reads: 8929, Coverage: 426, minDepth: 43\n",
      "Confirmed 3710 of 3888 bases (95.42%)\n",
      "Corrected 22 snps; 0 ambiguous bases; corrected 30 small insertions totaling 33 bases, 68 small deletions totaling 110 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000148:791-1103 0 -0 +0 NoSolution\n",
      "# fix break: tig00000148:2529-2738 0 -0 +0 NoSolution\n",
      "# fix break: tig00000148:2948-3013 0 -0 +0 NoSolution\n",
      "# fix break: tig00000148:3216-3663 0 -0 +0 NoSolution\n",
      "Finished processing tig00000148:1-3888\n",
      "Processing tig00000001:1-63791\n",
      "tig00000001:1-63791 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 0\n",
      "Total Reads: 160, Coverage: 0, minDepth: 5\n",
      "Confirmed 39 of 63791 bases (0.06%)\n",
      "Corrected 0 snps; 0 ambiguous bases; corrected 0 small insertions totaling 0 bases, 0 small deletions totaling 0 bases\n",
      "Finished processing tig00000001:1-63791\n",
      "Processing tig00000240:1-2508\n",
      "tig00000240:1-2508 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 327\n",
      "Total Reads: 4572, Coverage: 327, minDepth: 33\n",
      "Confirmed 2371 of 2508 bases (94.54%)\n",
      "Corrected 33 snps; 0 ambiguous bases; corrected 12 small insertions totaling 20 bases, 45 small deletions totaling 87 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000240:504 0 -0 +0 NoSolution\n",
      "# fix break: tig00000240:852-2026 0 -0 +0 NoSolution\n",
      "Finished processing tig00000240:1-2508\n",
      "Processing tig00000216:1-4319\n",
      "tig00000216:1-4319 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 327\n",
      "Total Reads: 7149, Coverage: 327, minDepth: 33\n",
      "Confirmed 4131 of 4319 bases (95.65%)\n",
      "Corrected 35 snps; 0 ambiguous bases; corrected 29 small insertions totaling 40 bases, 69 small deletions totaling 122 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000216:243-292 0 -0 +0 NoSolution\n",
      "fix break: tig00000216:582-1225 174 -1366 +1333 BreakFix\n",
      "# fix break: tig00000216:1501-1539 0 -0 +0 NoSolution\n",
      "# fix break: tig00000216:1828-2039 0 -0 +0 NoSolution\n",
      "# fix break: tig00000216:2262 0 -0 +0 NoSolution\n",
      "# fix break: tig00000216:2716-2894 0 -0 +0 NoSolution\n",
      "# fix break: tig00000216:3187-3254 0 -0 +0 NoSolution\n",
      "Finished processing tig00000216:1-4319\n",
      "Processing tig00000142:1-173\n",
      "tig00000142:1-173 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 222\n",
      "Total Reads: 888, Coverage: 222, minDepth: 22\n",
      "Confirmed 170 of 173 bases (98.27%)\n",
      "Corrected 2 snps; 0 ambiguous bases; corrected 0 small insertions totaling 0 bases, 1 small deletions totaling 1 bases\n",
      "Finished processing tig00000142:1-173\n",
      "Processing tig00000052:1-7658\n",
      "tig00000052:1-7658 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 253\n",
      "Total Reads: 9556, Coverage: 253, minDepth: 25\n",
      "Confirmed 7340 of 7658 bases (95.85%)\n",
      "Corrected 64 snps; 0 ambiguous bases; corrected 37 small insertions totaling 43 bases, 118 small deletions totaling 206 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000052:131-179 0 -0 +0 NoSolution\n",
      "fix break: tig00000052:457 88 -771 +750 BreakFix\n",
      "# fix break: tig00000052:1243-1407 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:1992-2090 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:2306 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:2627-2679 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:2958 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:3226-3294 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:3588-4834 0 -0 +0 NoSolution\n",
      "# fix break: tig00000052:5077 0 -0 +0 NoSolution\n",
      "fix break: tig00000052:5365-5681 4986 -1103 +1088 BreakFix\n",
      "fix break: tig00000052:6229-6511 5865 -932 +917 BreakFix\n",
      "# fix break: tig00000052:6796 0 -0 +0 NoSolution\n",
      "Finished processing tig00000052:1-7658\n",
      "Processing tig00000197:1-1483\n",
      "tig00000197:1-1483 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 343\n",
      "Total Reads: 3156, Coverage: 343, minDepth: 34\n",
      "Confirmed 1458 of 1483 bases (98.31%)\n",
      "Corrected 3 snps; 0 ambiguous bases; corrected 5 small insertions totaling 8 bases, 10 small deletions totaling 17 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "fix break: tig00000197:765-847 428 -754 +745 BreakFix\n",
      "Finished processing tig00000197:1-1483\n",
      "Processing tig00000077:1-2131\n",
      "tig00000077:1-2131 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 298\n",
      "Total Reads: 3566, Coverage: 298, minDepth: 30\n",
      "Confirmed 1969 of 2131 bases (92.40%)\n",
      "Corrected 40 snps; 0 ambiguous bases; corrected 21 small insertions totaling 37 bases, 51 small deletions totaling 98 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000077:483-952 0 -0 +0 NoSolution\n",
      "Finished processing tig00000077:1-2131\n",
      "Processing tig00001191:1-12031\n",
      "tig00000003:1-16669 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 1188\n",
      "Total Reads: 99842, Coverage: 1188, minDepth: 119\n",
      "Confirmed 15627 of 16669 bases (93.75%)\n",
      "Corrected 121 snps; 0 ambiguous bases; corrected 114 small insertions totaling 153 bases, 152 small deletions totaling 246 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000003:1594 0 -0 +0 NoSolution TandemRepeat 545\n",
      "# fix break: tig00000003:1814 0 -0 +0 NoSolution TandemRepeat 545\n",
      "# fix break: tig00000003:3203 0 -0 +0 NoSolution\n",
      "fix break: tig00000003:4478 4133 -648 +645 BreakFix\n",
      "# fix break: tig00000003:6090-6557 0 -0 +0 NoSolution\n",
      "fix break: tig00000003:8114 7786 -668 +656 BreakFix\n",
      "# fix break: tig00000003:12188-12261 0 -0 +0 NoSolution\n",
      "# fix break: tig00000003:12471-12779 0 -0 +0 NoSolution\n",
      "# fix break: tig00000003:14635-14676 0 -0 +0 NoSolution\n",
      "# fix break: tig00000003:14901-15020 0 -0 +0 NoSolution\n",
      "# fix break: tig00000003:15337-15996 0 -0 +0 NoSolution\n",
      "Finished processing tig00000003:1-16669\n",
      "Processing tig00000220:1-3021\n",
      "tig00000220:1-3021 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 265\n",
      "Total Reads: 4748, Coverage: 265, minDepth: 27\n",
      "Confirmed 2880 of 3021 bases (95.33%)\n",
      "Corrected 19 snps; 0 ambiguous bases; corrected 16 small insertions totaling 21 bases, 44 small deletions totaling 59 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000220:315-836 0 -0 +0 NoSolution\n",
      "fix break: tig00000220:1293-1370 1014 -694 +678 BreakFix\n",
      "# fix break: tig00000220:1573 0 -0 +0 NoSolution\n",
      "# fix break: tig00000220:1909-2052 0 -0 +0 NoSolution\n",
      "# fix break: tig00000220:2287 0 -0 +0 NoSolution\n",
      "Finished processing tig00000220:1-3021\n",
      "Processing tig00000059:1-766\n",
      "tig00000059:1-766 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 224\n",
      "Total Reads: 1783, Coverage: 224, minDepth: 22\n",
      "Confirmed 688 of 766 bases (89.82%)\n",
      "Corrected 11 snps; 0 ambiguous bases; corrected 3 small insertions totaling 3 bases, 17 small deletions totaling 28 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000059:222-412 0 -0 +0 NoSolution\n",
      "Finished processing tig00000059:1-766\n",
      "Processing tig00000179:1-476\n",
      "tig00000179:1-476 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 243\n",
      "Total Reads: 1580, Coverage: 243, minDepth: 24\n",
      "Confirmed 450 of 476 bases (94.54%)\n",
      "Corrected 8 snps; 0 ambiguous bases; corrected 2 small insertions totaling 2 bases, 10 small deletions totaling 14 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000179:280-364 0 -0 +0 NoSolution\n",
      "Finished processing tig00000179:1-476\n",
      "Processing tig00000191:1-1294\n",
      "tig00000191:1-1294 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 376\n",
      "Total Reads: 3273, Coverage: 376, minDepth: 38\n",
      "Confirmed 1255 of 1294 bases (96.99%)\n",
      "Corrected 5 snps; 0 ambiguous bases; corrected 6 small insertions totaling 9 bases, 15 small deletions totaling 26 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000191:429-491 0 -0 +0 NoSolution\n",
      "# fix break: tig00000191:974-1031 0 -0 +0 NoSolution\n",
      "Finished processing tig00000191:1-1294\n",
      "Processing tig00000015:1-362\n",
      "tig00000015:1-362 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 0\n",
      "Total Reads: 0, Coverage: 0, minDepth: 5\n",
      "Confirmed 0 of 362 bases (0.00%)\n",
      "Corrected 0 snps; 0 ambiguous bases; corrected 0 small insertions totaling 0 bases, 0 small deletions totaling 0 bases\n",
      "Finished processing tig00000015:1-362\n",
      "Processing tig00000018:1-7513\n",
      "tig00001191:1-12031 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 263\n",
      "Total Reads: 15525, Coverage: 263, minDepth: 26\n",
      "Confirmed 10943 of 12031 bases (90.96%)\n",
      "Corrected 81 snps; 0 ambiguous bases; corrected 78 small insertions totaling 98 bases, 149 small deletions totaling 245 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00001191:626-1250 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:1658-2339 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:2893-2895 0 -0 +0 NoSolution\n",
      "fix break: tig00001191:3260-3261 2857 -735 +723 BreakFix\n",
      "# fix break: tig00001191:3795-3796 0 -0 +0 NoSolution\n",
      "fix break: tig00001191:4082 3762 -709 +698 BreakFix\n",
      "# fix break: tig00001191:4467-4796 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:5010-5243 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:5570-5797 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:6126-6538 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:6840-7283 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:7514 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:7742-7800 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:8394 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:8718-8719 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:9565-9668 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:10157 0 -0 +0 NoSolution\n",
      "# fix break: tig00001191:10828-11327 0 -0 +0 NoSolution\n",
      "Finished processing tig00001191:1-12031\n",
      "Processing tig00000054:1-676\n",
      "tig00000054:1-676 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 115\n",
      "Total Reads: 862, Coverage: 115, minDepth: 12\n",
      "Confirmed 648 of 676 bases (95.86%)\n",
      "Corrected 8 snps; 0 ambiguous bases; corrected 5 small insertions totaling 6 bases, 7 small deletions totaling 15 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000054:143-248 0 -0 +0 NoSolution\n",
      "Finished processing tig00000054:1-676\n",
      "tig00000018:1-7513 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 292\n",
      "Total Reads: 10523, Coverage: 292, minDepth: 29\n",
      "Confirmed 7110 of 7513 bases (94.64%)\n",
      "Corrected 103 snps; 0 ambiguous bases; corrected 47 small insertions totaling 56 bases, 150 small deletions totaling 244 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000018:856-892 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:1106-1356 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:1895 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:2458-2653 0 -0 +0 NoSolution\n",
      "fix break: tig00000018:3034-3039 2657 -738 +716 BreakFix\n",
      "# fix break: tig00000018:3355-3358 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:3619-3986 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:4250-4708 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:4972-5391 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:5840 0 -0 +0 NoSolution\n",
      "fix break: tig00000018:6358-6360 6007 -707 +694 BreakFix\n",
      "# fix break: tig00000018:6635 0 -0 +0 NoSolution\n",
      "# fix break: tig00000018:6883-7232 0 -0 +0 NoSolution\n",
      "Finished processing tig00000018:1-7513\n",
      "Processing tig00000028:1-8135\n",
      "tig00000028:1-8135 log:\n",
      "frags data/sample/bwa_output/bwa_aligned_reads.bam: coverage 295\n",
      "Total Reads: 11722, Coverage: 295, minDepth: 30\n",
      "Confirmed 7765 of 8135 bases (95.45%)\n",
      "Corrected 89 snps; 1 ambiguous bases; corrected 63 small insertions totaling 84 bases, 119 small deletions totaling 213 bases\n",
      "# Attempting to fix local continuity breaks\n",
      "# fix break: tig00000028:663 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:878-995 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:1254-1390 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:1679-1706 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:1976-2560 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:2838-3346 0 -0 +0 NoSolution\n",
      "fix break: tig00000028:3621-3715 3239 -879 +850 BreakFix\n",
      "# fix break: tig00000028:4505-4697 0 -0 +0 NoSolution\n",
      "fix break: tig00000028:4905-5293 4505 -1184 +1156 BreakFix\n",
      "# fix break: tig00000028:5597-5598 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:5971-6030 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:6351-6672 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:6969-7494 0 -0 +0 NoSolution\n",
      "# fix break: tig00000028:7860-7862 0 -0 +0 NoSolution\n",
      "Finished processing tig00000028:1-8135\n",
      "Writing updated tig00000001_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000003_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000008_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000009_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000015_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000018_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000026_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000028_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000033_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000049_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000052_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000053_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000054_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000059_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000072_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000077_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000088_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000091_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000092_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000095_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000114_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000116_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000119_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000123_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000127_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000142_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000143_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000148_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000153_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000179_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000182_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000191_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000193_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000197_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000201_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000203_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000216_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000220_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000223_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000240_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000242_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000244_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000245_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00000250_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00001190_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Writing updated tig00001191_pilon to data/sample/pilon_output/pilon.contigs.fasta\n",
      "Mean frags coverage: 265\n",
      "Mean total coverage: 265\n"
     ]
    }
   ],
   "source": [
    "java -Xmx128g -XX:+UseConcMarkSweepGC \\\n",
    "      -XX:-UseGCOverheadLimit \\\n",
    "      -jar /home/jovyan/software/pilon/pilon-1.22.jar \\\n",
    "      --threads 2 \\\n",
    "      --genome  data/sample/racon_output/sample_racon.contigs.fasta \\\n",
    "      --bam data/sample/bwa_output/bwa_aligned_reads.bam \\\n",
    "      --outdir data/sample/pilon_output \\\n",
    "      --output pilon.contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *De novo* Miniasm assembly\n",
    " \n",
    "[Miniasm](https://github.com/lh3/miniasm) is a fast Overlap-Layout-Consensus-based option for *de novo* assembly of noisy long reads. Miniasm builds high-confidence contigs (unitigs) concatenating pieces of read sequences. It takes all read self-mappings as input to output an assembly graph. At least for high-coverage bacterial genomes, Miniasm can generate long contigs from raw ONT reads without error correction. The error rate of the assembly is the same as that of the raw input reads. In this way, Miniasm produces uncontaminated and uncorrected contig sequences from raw read overlays. To reduce the presence of artefacts such as adapters and untrimmed chimeras in the assembly, Miniasm calculates the per base coverage based on good mappings (longer than 2 kb with at least 100 bp non-redundant bases on matching minimizers) versus other reads. \n",
    "Furthermore, Miniasm ignores internal matches, eliminates contained reads, and adds overlays to the assembly graph. To avoid multiple edges, Miniasm uses the longest overlay. When the assembly graph is generated, it removes the transitive edges, trims the tilt units composed of few reads, and permits small bubbles to appear. \n",
    "\n",
    "Although it cannot produce a high-quality consensus, Miniasm is extremely fast and produces continuous and structurally accurate assemblies, at least for genomes without excessive repetitive sequences. \n",
    "\n",
    "Minimap is first used to get an all-vs-all read mappings of the reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::1.228*1.00] collected minimizers\n",
      "[M::mm_idx_gen::1.514*1.29] sorted minimizers\n",
      "[M::main::1.514*1.29] loaded/built the index for 3720 target sequence(s)\n",
      "[M::mm_mapopt_update::1.645*1.27] mid_occ = 24\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 3720\n",
      "[M::mm_idx_stat::1.738*1.26] distinct minimizers: 9912744 (88.47% are singletons); average occurrences: 1.197; average spacing: 2.937\n",
      "[M::worker_pipeline::4.286*2.03] mapped 3720 sequences\n",
      "[M::main] Version: 2.14-r892-dirty\n",
      "[M::main] CMD: minimap2 -x ava-ont data/sample/reads.fastq data/sample/reads.fastq\n",
      "[M::main] Real time: 4.314 sec; CPU: 8.732 sec; Peak RSS: 0.528 GB\n"
     ]
    }
   ],
   "source": [
    "minimap2 -x ava-ont data/sample/reads.fastq data/sample/reads.fastq | gzip -1 > data/sample/reads.paf.gz "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Miniasm takes the obtained mappings file and the FASTA reads file and generates the final assembly. The assembly algorithm doesn't have a consensus step as it usually needs multiple steps to produce a precise consensus sequence, and that is a computational bottleneck."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::main] ===> Step 1: reading read mappings <===\n",
      "[M::ma_hit_read::0.016*0.74] read 11398 hits; stored 11523 hits and 757 sequences (7765831 bp)\n",
      "[M::main] ===> Step 2: 1-pass (crude) read selection <===\n",
      "[M::ma_hit_sub::0.017*0.94] 413 query sequences remain after sub\n",
      "[M::ma_hit_cut::0.017*0.92] 7892 hits remain after cut\n",
      "[M::ma_hit_flt::0.017*0.92] 7846 hits remain after filtering; crude coverage after filtering: 18.93\n",
      "[M::main] ===> Step 3: 2-pass (fine) read selection <===\n",
      "[M::ma_hit_sub::0.018*0.89] 308 query sequences remain after sub\n",
      "[M::ma_hit_cut::0.018*0.89] 7486 hits remain after cut\n",
      "[M::ma_hit_contained::0.018*0.88] 32 sequences and 28 hits remain after containment removal\n",
      "[M::main] ===> Step 4: graph cleaning <===\n",
      "[M::ma_sg_gen] read 28 arcs\n",
      "[M::main] ===> Step 4.1: transitive reduction <===\n",
      "[M::asg_arc_del_trans] transitively reduced 8 arcs\n",
      "[M::asg_arc_del_multi] removed 0 multi-arcs\n",
      "[M::asg_arc_del_asymm] removed 0 asymmetric arcs\n",
      "[M::main] ===> Step 4.2: initial tip cutting and bubble popping <===\n",
      "[M::asg_cut_tip] cut 21 tips\n",
      "[M::asg_pop_bubble] popped 0 bubbles and trimmed 0 tips\n",
      "[M::main] ===> Step 4.3: cutting short overlaps (3 rounds in total) <===\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::main] ===> Step 4.4: removing short internal sequences and bi-loops <===\n",
      "[M::asg_cut_internal] cut 0 internal sequences\n",
      "[M::asg_cut_biloop] cut 0 small bi-loops\n",
      "[M::asg_cut_tip] cut 0 tips\n",
      "[M::asg_pop_bubble] popped 0 bubbles and trimmed 0 tips\n",
      "[M::main] ===> Step 4.5: aggressively cutting short overlaps <===\n",
      "[M::asg_arc_del_short] removed 0 short overlaps\n",
      "[M::main] ===> Step 5: generating unitigs <===\n",
      "[M::main] Version: 0.3-r179\n",
      "[M::main] CMD: miniasm -f data/sample/reads.fastq data/sample/reads.paf.gz\n",
      "[M::main] Real time: 0.089 sec; CPU: 0.076 sec\n"
     ]
    }
   ],
   "source": [
    "mkdir -p data/sample/miniasm_output\n",
    "miniasm -f data/sample/reads.fastq data/sample/reads.paf.gz  > data/sample/miniasm_output/assembly_with_miniasm.gfa "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "\n",
    "[1] Loman N.J., Quick J. and Simpson J.T. A complete bacterial genome assembled de novo using only nanopore sequencing data. Nature Methods 2015 12:733–735. DOI https://doi.org/10.1101/015552\n",
    "\n",
    "[2] Li H. Minimap and miniasm: fast mapping and de novo assembly for noisy long sequences. Bioinformatics, Volume 32, Issue 14, 15 July 2016, Pages 2103–2110. DOI https://doi.org/10.1093/bioinformatics/btw152"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
