{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hybrid assemblies\n",
    "\n",
    "##### *NOTE: This notebook has high memory requirements*\n",
    "\n",
    "Obtaining the assembled sequence of a complete genome is a complex multi-step task. For *De novo* assembly, the simplest elements of the hierarchy are the reads provided after the sequencing. The next level of hierarchy is the alignment of multiple reads without definite order (i.e. contigs). Finally, the top level of the hierarchy corresponds to the sum of two or more contigs where a (near) complete structure of the genome under study is obtained. Ideally, one expects to obtain a single fragment (contig) for each chromosome or a plasmid that is present in the genome. However, most of the times the assemblies are incomplete, especially when dealing with short reads that can be caused by the process of preparing the material or by the technological limitations. Specifically, when a repeat region is longer than the reads, this will create a single contig in the assembly with multiple connections. ONT technologies provide long reads that helps to build bridges between the contigs generated by short reads, improving the continuity of the assemblies. To put it simple, the hybrid assemblies use short reads to produce accurate contigs and the long reads to provide information to assemble them together."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hybrid assembly with Unicycler\n",
    "\n",
    "Unicycler is a hybrid assembly for circular genomes based on a combination of short (Illumina) and long (ONT) reads, which produces assemblies that are accurate and complete (using depth and connectivity information). Unicycler has a very low misassembly rate. However, it is not particularly fast, with necessary time being dependent on the number of long reads and the genome size/complexity. It can use long reads of any depth and quality. The recommendations suggest a coverage of more than 10× in long reads to be able to warrant completing a genome. Despite this, Unicycler can assembly nearly-complete genomes with fewer long reads as long as an optimal coverage with short reads is attained. Furthermore, it is very easy to use and produce an assembly graph. \n",
    "\n",
    "Unicycler relies on SPAdes to perform the assembly, which proceeds using De Bruijn graphs with a range of different sizes of k-mer, and then does multiple rounds of short-read polishing using Pilon. Unicycler eliminates contigs with a depth of less than half the depth of the average graph, avoiding possible contaminations. It also eliminates bridges that may be wrong, assigning a quality score to each bridge and applying them in decreasing order. The quality of the bridge depends on the number of reads that support the bridge, the quality of alignment between the read consensus and the path of the graph, the length of the two contigs to be bridged, the length and quality of the read alignments to the contigs, and the consistency of the read between the contigs. An ideal long read bridge, therefore, connects two long contigs with the same depth.\n",
    "\n",
    "Unicycler uses SeqAn as a semi-global aligner to produce a consensus break sequence and find the best graphic path that connects the contigs, through a branch and boundary algorithm. If any circular replicon was completely assembled, now it will be a single contig with a link that will connect its end to its beginning. In this case, Unicycler uses TBLASTN to search for dnaA or repA alleles in each completed replicon which provides consistently oriented assemblies and reduces the risk of a gene dividing at the beginning and the end of the sequence. As a final step, Unicycler uses Bowtie2 and Pilon to polish the assembly and reduce the rate of small errors.\n",
    "\n",
    "Unicycler can be run in three different modes: Conservative, where only very high-quality bridges are used; Bold, which use inferior quality bridges; and Normal mode, which uses an intermediate quality and is recommended in most cases.\n",
    "\n",
    "Short Illumina/IonTorrent read sets can be combined with ONT read sets. For that, the Unicycler basic command performs both assembly and polish tasks and takes the following inputs:\n",
    "\n",
    "- **-1 and -2**: Illumina reads.\n",
    "- **-l**: Long reads (ONT in our case).\n",
    "- **-s**: IonTorrent reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### REMARK: **Data should not be included in /data folder of this project (see \"Additional data\" section in README file)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Unicycler\u001b[0m (2018-12-02 00:26:41)\n",
      "    Welcome to Unicycler, an assembly pipeline for bacterial genomes. Since you\n",
      "provided both short and long reads, Unicycler will perform a hybrid assembly.\n",
      "It will first use SPAdes to make a short-read assembly graph, and then it will\n",
      "use various methods to scaffold that graph with the long reads.\n",
      "    For more information, please see https://github.com/rrwick/Unicycler\n",
      "\n",
      "Command: \u001b[1m/home/jovyan/software/Unicycler/unicycler-runner.py -1 data/sample/short/reads_1.fastq.gz -2 data/sample/short/reads_2.fastq.gz -l data/sample/reads.fastq -o data/sample/hybrid_output\u001b[0m\n",
      "\n",
      "Unicycler version: v0.4.8-beta\n",
      "Using 3 threads\n",
      "\n",
      "The output directory already exists and files may be reused or overwritten:\n",
      "  /home/jovyan/notebooks/data/sample/hybrid_output\n",
      "\n",
      "Dependencies:\n",
      "  \u001b[4mProgram         Version     Status  \u001b[0m\n",
      "  spades.py       3.13.0      \u001b[32mgood\u001b[0m    \n",
      "  racon           1.3.1       \u001b[32mgood\u001b[0m    \n",
      "  makeblastdb     2.7.1+      \u001b[32mgood\u001b[0m    \n",
      "  tblastn         2.7.1+      \u001b[32mgood\u001b[0m    \n",
      "  bowtie2-build   2.3.4.1     \u001b[32mgood\u001b[0m    \n",
      "  bowtie2         2.3.4.1     \u001b[32mgood\u001b[0m    \n",
      "  samtools        1.5         \u001b[32mgood\u001b[0m    \n",
      "  java            1.8.0_191   \u001b[32mgood\u001b[0m    \n",
      "  pilon           1.22        \u001b[32mgood\u001b[0m    \n",
      "  bcftools                    not used\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mSPAdes read error correction\u001b[0m (2018-12-02 00:27:07)\n",
      "    Unicycler uses the SPAdes read error correction module to reduce the number\n",
      "of errors in the short read before SPAdes assembly. This can make the assembly\n",
      "faster and simplify the assembly graph structure.\n",
      "\n",
      "Command: \u001b[1m/home/jovyan/software/SPAdes-3.13.0-Linux/bin/spades.py -1 /home/jovyan/notebooks/data/sample/short/reads_1.fastq.gz -2 /home/jovyan/notebooks/data/sample/short/reads_2.fastq.gz -o /home/jovyan/notebooks/data/sample/hybrid_output/spades_assembly/read_correction --threads 3 --only-error-correction\u001b[0m\n",
      "\n",
      "Corrected reads:\n",
      "  /home/jovyan/notebooks/data/sample/hybrid_output/spades_assembly/corrected_1.fastq.gz\n",
      "  /home/jovyan/notebooks/data/sample/hybrid_output/spades_assembly/corrected_2.fastq.gz\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mChoosing k-mer range for assembly\u001b[0m (2018-12-02 00:54:21)\n",
      "    Unicycler chooses a k-mer range for SPAdes based on the length of the input\n",
      "reads. It uses a wide range of many k-mer sizes to maximise the chance of\n",
      "finding an ideal assembly.\n",
      "\n",
      "SPAdes maximum k-mer: 127\n",
      "Median read length: 251\n",
      "K-mer range: 27, 47, 63, 77, 89, 99, 107, 115, 121, 127\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mSPAdes assemblies\u001b[0m (2018-12-02 00:54:44)\n",
      "    Unicycler now uses SPAdes to assemble the short reads. It scores the\n",
      "assembly graph for each k-mer using the number of contigs (fewer is better) and\n",
      "the number of dead ends (fewer is better). The score function is 1/(c*(d+2)),\n",
      "where c is the contig count and d is the dead end count.\n",
      "\n",
      "\u001b[4mK-mer   Contigs   Dead ends   Score   \u001b[0m\n",
      "   27       421           4   3.96e-04\n",
      "   47       154           2   1.62e-03\n",
      "   63       119           2   2.10e-03\n",
      "   77        98           3   2.04e-03\n",
      "   89        81           1   4.12e-03\n",
      "   99        76           0   6.58e-03\n",
      "\u001b[32m  107        73           0   6.85e-03 ← best\u001b[0m\n",
      "  115        73           0   6.85e-03\n",
      "  121        73           0   6.85e-03\n",
      "  127        73           0   6.85e-03\n",
      "\n",
      "Deleting /home/jovyan/notebooks/data/sample/hybrid_output/spades_assembly/\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mDetermining graph multiplicity\u001b[0m (2018-12-02 01:14:46)\n",
      "    Multiplicity is the number of times a sequence occurs in the underlying\n",
      "sequence. Single-copy contigs (those with a multiplicity of one, occurring only\n",
      "once in the underlying sequence) are particularly useful.\n",
      "\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/001_best_spades_graph.gfa\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mCleaning graph\u001b[0m (2018-12-02 01:14:46)\n",
      "    Unicycler now performs various cleaning procedures on the graph to remove\n",
      "overlaps and simplify the graph structure. The end result is a graph ready for\n",
      "bridging.\n",
      "\n",
      "Graph overlaps removed\n",
      "\n",
      "Removed zero-length segments:\n",
      "    49, 50, 51, 52, 56, 58, 59\n",
      "\n",
      "Removed zero-length segments:\n",
      "    67\n",
      "\n",
      "Merged small segments:\n",
      "    69, 70, 71, 72, 73\n",
      "\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/002_overlaps_removed.gfa\n",
      "\n",
      "    Unicycler now selects a set of anchor contigs from the single-copy contigs.\n",
      "These are the contigs which will be connected via bridges to form the final\n",
      "assembly.\n",
      "\n",
      "24 anchor segments (2,132,850 bp) out of 60 total segments (2,147,208 bp)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mCreating SPAdes contig bridges\u001b[0m (2018-12-02 01:14:46)\n",
      "    SPAdes uses paired-end information to perform repeat resolution (RR) and\n",
      "produce contigs from the assembly graph. SPAdes saves the graph paths\n",
      "corresponding to these contigs in the contigs.paths file. When one of these\n",
      "paths contains two or more anchor contigs, Unicycler can create a bridge from\n",
      "the path.\n",
      "\n",
      "                                Bridge\u001b[0m\n",
      "\u001b[4mStart       Path       End     quality\u001b[0m\n",
      "    1     29 → 29      3          19.8\n",
      "    7   49 → 44 → 46   -8         63.2\n",
      "   11        40        22         37.2\n",
      "   16   49 → 44 → 46   -17        58.8\n",
      "   22       -34        19         52.8\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mCreating loop unrolling bridges\u001b[0m (2018-12-02 01:14:46)\n",
      "    When a SPAdes contig path connects an anchor contig with the middle contig\n",
      "of a simple loop, Unicycler concludes that the sequences are contiguous (i.e.\n",
      "the loop is not a separate piece of DNA). It then uses the read depth of the\n",
      "middle and repeat contigs to guess the number of times to traverse the loop and\n",
      "makes a bridge.\n",
      "\n",
      "No loop unrolling bridges made\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mLoading reads\u001b[0m (2018-12-02 01:14:46)\n",
      "3,720 / 3,720 (100.0%) - 34,842,223 bp\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mAssembling contigs and long reads with miniasm\u001b[0m (2018-12-02 01:14:47)\n",
      "    Unicycler uses miniasm to construct a string graph assembly using both the\n",
      "short read contigs and the long reads. It will then use the resulting string\n",
      "graph to produce bridges between contigs. This method requires decent coverage\n",
      "of long reads and therefore may not be fruitful if long reads are sparse.\n",
      "However, it does not rely on the short read assembly graph having good\n",
      "connectivity and is able to bridge an assembly graph even when it contains many\n",
      "dead ends.\n",
      "    Unicycler uses two types of \"reads\" as assembly input: anchor contigs from\n",
      "the short-read assembly and actual long reads which overlap two or more of\n",
      "these contigs. It then assembles them with miniasm.\n",
      "\n",
      "Aligning long reads to graph using minimap\n",
      "\n",
      "Saving to /home/jovyan/notebooks/data/sample/hybrid_output/miniasm_assembly/01_assembly_reads.fastq:\n",
      "  23 short-read contigs\n",
      "  148 long reads\n",
      "\n",
      "Finding overlaps with minimap... \u001b[32msuccess\u001b[0m\n",
      "  1,235 overlaps\n",
      "\n",
      "Assembling reads with miniasm... \u001b[32msuccess\u001b[0m\n",
      "  6 segments, 5 links\n",
      "\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/miniasm_assembly/11_branching_paths_removed.gfa\n",
      "Merging segments into unitigs:\n",
      "  1 linear unitig\n",
      "  total size = 524,904 bp\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/miniasm_assembly/12_unitig_graph.gfa\n",
      "\n",
      "\u001b[31mminiasm assembly too small for bridging\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mCreating simple long read bridges\u001b[0m (2018-12-02 01:14:49)\n",
      "    Unicycler uses long read alignments (from minimap) to resolve simple repeat\n",
      "structures in the graph. This takes care of some \"low-hanging fruit\" of the\n",
      "graph simplification.\n",
      "\n",
      "Aligning long reads to graph using minimap\n",
      "\n",
      "    Two-way junctions are defined as cases where two graph contigs (A and B)\n",
      "join together (C) and then split apart again (D and E). This usually represents\n",
      "a simple 2-copy repeat, and there are two possible options for its resolution:\n",
      "(A→C→D and B→C→E) or (A→C→E and B→C→D). Each read which spans such a junction\n",
      "gets to \"vote\" for option 1, option 2 or neither. Unicycler creates a bridge at\n",
      "each junction for the most voted for option.\n",
      "\n",
      "                                            Op. 1   Op. 2   Neither   Final    Bridge\u001b[0m\n",
      "\u001b[4mJunction      Option 1        Option 2      votes   votes     votes     op.   quality\u001b[0m\n",
      "      28   5 → 28 → -15,    5 → 28 → 18,        2       0         8       1       0.0\n",
      "            20 → 28 → 18   20 → 28 → -15                                             \n",
      "\n",
      "    Simple loops are parts of the graph where two contigs (A and B) are\n",
      "connected via a repeat (C) which loops back to itself (via D). It is possible\n",
      "to traverse the loop zero times (A→C→B), one time (A→C→D→C→B), two times\n",
      "(A→C→D→C→D→C→B), etc. Long reads which span the loop inform which is the\n",
      "correct number of times through. In this step, such reads are found and each is\n",
      "aligned against alternative loop counts. A reads casts its \"vote\" for the loop\n",
      "count it agrees best with, and Unicycler creates a bridge using the most voted\n",
      "for count.\n",
      "\n",
      "                                   Read                         Loop    Bridge\u001b[0m\n",
      "\u001b[4mStart   Repeat   Middle     End   count   Read votes           count   quality\u001b[0m\n",
      "  -21       38                7       1   5 loops: 1 vote          5      70.2\n",
      "    1       29                3       6   bad: 1 vote    1         1      76.4\n",
      "                                          loop: 5 votes                       \n",
      "   11       40               22       3   bad: 1 vote    1         1      43.5\n",
      "                                          loop: 2 votes                       \n",
      "   19       37                2       1   1 loop: 1 vote           1      69.2\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mDetermining low score threshold\u001b[0m (2018-12-02 01:14:57)\n",
      "    Before conducting semi-global alignment of the long reads to the assembly\n",
      "graph, Unicycler must determine a minimum alignment score threshold such that\n",
      "nonsense alignments are excluded. To choose a threshold automatically, it\n",
      "examines alignments between random sequences and selects a score a few standard\n",
      "deviations above the mean.\n",
      "\n",
      "Automatically choosing a threshold using random alignment scores.\n",
      "\n",
      "Random alignment mean score: 61.66\n",
      "         standard deviation:  1.31\n",
      "        Low score threshold: 61.66 + (7 x 1.31) = 70.86\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mAligning reads\u001b[0m (2018-12-02 01:14:57)\n",
      "3,720 / 3,720 (100.0%)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mRead alignment summary\u001b[0m (2018-12-02 01:16:12)\n",
      "Total read count:            3,720\n",
      "Fully aligned reads:         1,177\n",
      "Partially aligned reads:        62\n",
      "Unaligned reads:             2,481\n",
      "Total bases aligned:     5,502,777 bp\n",
      "Mean alignment identity:        84.5%\n",
      "\n",
      "Deleting /home/jovyan/notebooks/data/sample/hybrid_output/read_alignment/\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mBuilding long read bridges\u001b[0m (2018-12-02 01:16:12)\n",
      "    Unicycler uses the long read alignments to produce bridges between anchor\n",
      "segments. These bridges can be formed using as few as one long read, giving\n",
      "Unicycler the ability to bridge the graph even when long-read depth is low.\n",
      "\n",
      "\u001b[4m       Start → end    Best path                                 Quality\u001b[0m\n",
      " 1/14     -7 → 21     -38, -38, -38, -38, -38, -38               89.621\n",
      " 2/14     23 → -3     -27                                        84.396\n",
      " 3/14     -8 → 16     56, 39, 57, 39, 56, 39, 57, 39             50.526\n",
      " 4/14     20 → 18     28                                         64.006\n",
      " 5/14     -6 → 20     30, 34                                     46.060\n",
      " 6/14     18 → 10     -34, -30                                   72.286\n",
      " 7/14      6 → 1      -34, -30                                   52.152\n",
      " 8/14      2 → 17     -34, -30                                   57.818\n",
      " 9/14     19 → 2      37, 37                                     60.017\n",
      "10/14     17 → -16    -46, -44, -49                              74.484\n",
      "11/14      1 → 3      29, 29                                     73.070\n",
      "12/14     22 → 19     -34                                        44.659\n",
      "13/14     11 → 22     40, 40                                     78.582\n",
      "14/14      9 → -4     -31, -36, -43, -48, -51, 24, -52, 54,      65.126\n",
      "                      53, -60                                          \n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mApplying bridges\u001b[0m (2018-12-02 01:16:12)\n",
      "    Unicycler now applies to the graph in decreasing order of quality. This\n",
      "ensures that when multiple, contradictory bridges exist, the most supported\n",
      "option is used.\n",
      "\n",
      "\u001b[4mBridge type        Start → end   Path                                       Quality\u001b[0m\n",
      "long read             -7 → 21    -38, -38, -38, -38, -38, -38                89.621\n",
      "long read             23 → -3    -27                                         84.396\n",
      "long read             11 → 22    40, 40                                      78.582\n",
      "simple long read       1 → 3     29, 29                                      76.416\n",
      "long read             17 → -16   -46, -44, -49                               74.484\n",
      "long read             18 → 10    -34, -30                                    72.286\n",
      "simple long read      19 → 2     37, 37                                      69.247\n",
      "long read              9 → -4    -31, -36, -43, -48, -51, 24, -52, 54,       65.126\n",
      "                                 53, -60                                           \n",
      "long read             20 → 18    28                                          64.006\n",
      "long read              2 → 17    -34, -30                                    57.818\n",
      "long read              6 → 1     -34, -30                                    52.152\n",
      "long read             -8 → 16    56, 39, 57, 39, 56, 39, 57, 39              50.526\n",
      "long read             -6 → 20    30, 34                                      46.060\n",
      "long read             22 → 19    -34                                         44.659\n",
      "SPAdes                 7 → -8    49, 44, 46                                  63.155\n",
      "\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/003_bridges_applied.gfa\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mBridged assembly graph\u001b[0m (2018-12-02 01:16:13)\n",
      "    The assembly is now mostly finished and no more structural changes will be\n",
      "made. Ideally the assembly graph should now have one contig per replicon and no\n",
      "erroneous contigs (i.e a complete assembly). If there are more contigs, then\n",
      "the assembly is not complete.\n",
      "\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/004_final_clean.gfa\n",
      "\n",
      "\u001b[4mComponent   Segments   Links   Length      N50       Longest segment   Status    \u001b[0m\n",
      "    total         24      32   2,161,723   714,293         1,019,216             \n",
      "        1         23      31   2,159,232   714,293         1,019,216  \u001b[31m incomplete\u001b[0m\n",
      "        2          1       1       2,491     2,491             2,491    \u001b[32m complete\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mPolishing assembly with Pilon\u001b[0m (2018-12-02 01:16:13)\n",
      "    Unicycler now conducts multiple rounds of Pilon in an attempt to repair any\n",
      "remaining small-scale errors with the assembly.\n",
      "\n",
      "Aligning reads to find appropriate insert size range...\n",
      "Insert size 1st percentile:  257\n",
      "Insert size 99th percentile: 1026\n",
      "\n",
      "\u001b[4mPilon polish round 1\u001b[0m\n",
      "Total number of changes: 10\n",
      "\n",
      "\u001b[4mPilon polish round 2\u001b[0m\n",
      "No Pilon changes\n",
      "\n",
      "\u001b[4mPilon polish round 3\u001b[0m\n",
      "No Pilon changes\n",
      "\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/005_polished.gfa\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mRotating completed replicons\u001b[0m (2018-12-02 02:49:32)\n",
      "    Any completed circular contigs (i.e. single contigs which have one link\n",
      "connecting end to start) can have their start position changed with altering\n",
      "the sequence. For consistency, Unicycler now searches for a starting gene (dnaA\n",
      "or repA) in each such contig, and if one is found, the contig is rotated to\n",
      "start with that gene on the forward strand.\n",
      "\n",
      "\u001b[4mSegment   Length   Depth   Starting gene   Position   Strand   Identity   Coverage\u001b[0m\n",
      "      8    2,491   0.29x   \u001b[31mnone found\u001b[0m                                             \n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mAssembly complete\u001b[0m (2018-12-02 02:49:39)\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/assembly.gfa\n",
      "Saving /home/jovyan/notebooks/data/sample/hybrid_output/assembly.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Data not included in repository\n",
    "unicycler-runner.py \\\n",
    "                      -1 data/sample/short/reads_1.fastq.gz \\\n",
    "                      -2 data/sample/short/reads_2.fastq.gz \\\n",
    "                      -l data/sample/reads.fastq \\\n",
    "                      -o data/sample/hybrid_output   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hybrid assembly with MaSuRCA\n",
    "\n",
    "[MaSuRCA](http://www.genome.umd.edu/masurca.html) is a *de novo* assembler that has the capacity to assemble both short reads only or a mixture of short (Illumina) and long reads (ONT) of animal or plant genomes. It combines the efficiency of the De Bruijn graphs (DBG) and the Overlap-Layout-Consensus (OLC) approaches. With OLC, it attempts to calculate all overlaps by pairs between reads, using the sequence similarity to determine the superposition. While with DBG, it represents the overlaps between the sequences using k-mers, avoiding the calculation of the overlap by pairs. The computational requirements vary with the size of the genome to be assembled. The memory usage scales linearly with the size of the genome, and the execution time scales linearly with the depth of coverage.\n",
    "\n",
    "MaSuRCA uses a modified version of the CABOG assembler, for the superposition-based assembly after the construction of super reads. The basic concept of super reads is to extend each original read forward and backward, base by base. Each of the original reads is contained in a super read and many of the original reads produce the same super read, so it leaves a very small data set to use.\n",
    "\n",
    "MaSuRCA can use BioNano Genomics (BNG) maps to improve the assembly of several highly repetitive genomes, where distances and positions of the same restriction sites are compared to a restriction map built computationally. The BNG map can detect rearrangements, insertions or eliminations on a large scale. The assembly algorithms can find the correct location for these elements if the input includes a track that is long enough to contain the full range of one repetition plus unique flank regions on each side. The last important step in MaSuRCA is the filling of spaces in the scaffolds that are relatively short and do not contain complicated repetitive structures. \n",
    "\n",
    "MaSuRCA has a configuration file where arguments and input data paths are placed. Once the configuration file is ready, the \"masurca\" binary script will check the configuration file and load the arguments. In this example, the configuration file is created only providing the necessary arguments to build an assembly from Illumina short reads and ONT long reads. MaSuRCA assemblies are better if several PE libraries are used. Finally, the \"assemble.sh\" script is executed to build the hybrid assembly.\n",
    "\n",
    "The config file consists of two sections: DATA and PARAMETERS. Each section concludes with END statement. User should copy the sample config file to the directory of choice for running the assembly and then modify it according to the specifications of the assembly project.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#The following command builds the configuration file adding the lines below from \"DATA\" to the \"EOL\" lines in an empty file.\n",
    "mkdir -p data/sample/masurca_output\n",
    "cd data/sample/masurca_output\n",
    "cat > config_file.txt << EOL\n",
    "DATA\n",
    "PE= pe 300 15  ../short/reads_1.fastq.gz ../short/reads_2.fastq.gz\n",
    "NANOPORE=../reads.fastq\n",
    "END\n",
    "PARAMETERS\n",
    "GRAPH_KMER_SIZE = auto\n",
    "USE_LINKING_MATES = 0\n",
    "LIMIT_JUMP_COVERAGE = 300\n",
    "CA_PARAMETERS =  cgwErrorRate=0.15\n",
    "KMER_COUNT_THRESHOLD = 1\n",
    "NUM_THREADS = 48\n",
    "JF_SIZE = 63000000\n",
    "END\n",
    "EOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Verifying PATHS...\n",
      "jellyfish OK\n",
      "runCA OK\n",
      "createSuperReadsForDirectory.perl OK\n",
      "creating script file for the actions...done.\n",
      "execute assemble.sh to run assembly\n",
      "\u001b[0;32m[Mon Dec  3 21:17:06 UTC 2018]\u001b[0m Processing pe library reads\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\u001b[0;32m[Mon Dec  3 21:17:23 UTC 2018]\u001b[0m Average PE read length 243\n",
      "\u001b[0;32m[Mon Dec  3 21:17:24 UTC 2018]\u001b[0m Using kmer size of 127 for the graph\n",
      "cat: write error: Broken pipe\n",
      "\u001b[0;32m[Mon Dec  3 21:17:24 UTC 2018]\u001b[0m MIN_Q_CHAR: 33\n",
      "\u001b[0;32m[Mon Dec  3 21:17:24 UTC 2018]\u001b[0m Creating mer database for Quorum\n",
      "\u001b[0;32m[Mon Dec  3 21:18:05 UTC 2018]\u001b[0m Error correct PE\n",
      "\u001b[0;32m[Mon Dec  3 21:24:10 UTC 2018]\u001b[0m Estimating genome size\n",
      "\u001b[0;32m[Mon Dec  3 21:24:51 UTC 2018]\u001b[0m Estimated genome size: 3702060\n",
      "\u001b[0;32m[Mon Dec  3 21:24:51 UTC 2018]\u001b[0m Creating k-unitigs with k=127\n",
      "\u001b[0;32m[Mon Dec  3 21:28:05 UTC 2018]\u001b[0m Computing super reads from PE \n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Using CABOG from /home/jovyan/software/MaSuRCA-3.2.8/bin/../CA8/Linux-amd64/bin\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Running mega-reads correction/assembly\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Using mer size 15 for mapping, B=15, d=0.02\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Estimated Genome Size 3702060\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Estimated Ploidy 1\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Using 48 threads\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Output prefix mr.41.15.15.0.02\n",
      "\n",
      "gzip: stdout: Broken pipe\n",
      "\u001b[0;32m[Mon Dec  3 21:28:24 UTC 2018]\u001b[0m Using 25x of the longest ONT reads\n",
      "\u001b[0;32m[Mon Dec  3 21:28:25 UTC 2018]\u001b[0m Reducing super-read k-mer size\n",
      "\u001b[0;32m[Mon Dec  3 21:28:49 UTC 2018]\u001b[0m Mega-reads pass 1\n",
      "\u001b[0;32m[Mon Dec  3 21:28:49 UTC 2018]\u001b[0m Running locally in 1 batch\n",
      "\u001b[0;32m[Mon Dec  3 21:28:57 UTC 2018]\u001b[0m Mega-reads pass 2\n",
      "\u001b[0;32m[Mon Dec  3 21:28:57 UTC 2018]\u001b[0m Running locally in 1 batch\n",
      "\u001b[0;32m[Mon Dec  3 21:29:02 UTC 2018]\u001b[0m Refining alignments\n",
      "\u001b[0;32m[Mon Dec  3 21:29:07 UTC 2018]\u001b[0m Joining\n",
      "\u001b[0;32m[Mon Dec  3 21:29:07 UTC 2018]\u001b[0m Gap consensus\n",
      "xargs: invalid number \"-I\" for -P option\n",
      "Usage: xargs [OPTION]... COMMAND [INITIAL-ARGS]...\n",
      "Run COMMAND with arguments INITIAL-ARGS and more arguments read from input.\n",
      "\n",
      "Mandatory and optional arguments to long options are also\n",
      "mandatory or optional for the corresponding short option.\n",
      "  -0, --null                   items are separated by a null, not whitespace;\n",
      "                                 disables quote and backslash processing and\n",
      "                                 logical EOF processing\n",
      "  -a, --arg-file=FILE          read arguments from FILE, not standard input\n",
      "  -d, --delimiter=CHARACTER    items in input stream are separated by CHARACTER,\n",
      "                                 not by whitespace; disables quote and backslash\n",
      "                                 processing and logical EOF processing\n",
      "  -E END                       set logical EOF string; if END occurs as a line\n",
      "                                 of input, the rest of the input is ignored\n",
      "                                 (ignored if -0 or -d was specified)\n",
      "  -e, --eof[=END]              equivalent to -E END if END is specified;\n",
      "                                 otherwise, there is no end-of-file string\n",
      "  -I R                         same as --replace=R\n",
      "  -i, --replace[=R]            replace R in INITIAL-ARGS with names read\n",
      "                                 from standard input; if R is unspecified,\n",
      "                                 assume {}\n",
      "  -L, --max-lines=MAX-LINES    use at most MAX-LINES non-blank input lines per\n",
      "                                 command line\n",
      "  -l[MAX-LINES]                similar to -L but defaults to at most one non-\n",
      "                                 blank input line if MAX-LINES is not specified\n",
      "  -n, --max-args=MAX-ARGS      use at most MAX-ARGS arguments per command line\n",
      "  -P, --max-procs=MAX-PROCS    run at most MAX-PROCS processes at a time\n",
      "  -p, --interactive            prompt before running commands\n",
      "      --process-slot-var=VAR   set environment variable VAR in child processes\n",
      "  -r, --no-run-if-empty        if there are no arguments, then do not run COMMAND;\n",
      "                                 if this option is not given, COMMAND will be\n",
      "                                 run at least once\n",
      "  -s, --max-chars=MAX-CHARS    limit length of command line to MAX-CHARS\n",
      "      --show-limits            show limits on command-line length\n",
      "  -t, --verbose                print commands before executing them\n",
      "  -x, --exit                   exit if the size (see -s) is exceeded\n",
      "      --help                   display this help and exit\n",
      "      --version                output version information and exit\n",
      "\n",
      "Please see also the documentation at http://www.gnu.org/software/findutils/.\n",
      "You can report (and track progress on fixing) bugs in the \"xargs\"\n",
      "program via the GNU findutils bug-reporting page at\n",
      "https://savannah.gnu.org/bugs/?group=findutils or, if\n",
      "you have no web access, by sending email to <bug-findutils@gnu.org>.\n",
      "xargs: ./do_consensus.sh: No such file or directory\n",
      "cat: 'merges.[0-9]*.txt': No such file or directory\n",
      "\u001b[0;32m[Mon Dec  3 21:29:07 UTC 2018]\u001b[0m Warning! Some or all gap consensus jobs failed, see files in mr.41.15.15.0.02.join_consensus.tmp, proceeding anyway, to rerun gap consensus erase mr.41.15.15.0.02.1.fa and re-run assemble.sh\n",
      "\u001b[0;32m[Mon Dec  3 21:29:07 UTC 2018]\u001b[0m Generating assembly input files\n",
      "\u001b[0;32m[Mon Dec  3 21:29:19 UTC 2018]\u001b[0m Coverage of the mega-reads less than 5 -- using the super reads as well\n",
      "\u001b[0;32m[Mon Dec  3 21:29:20 UTC 2018]\u001b[0m Coverage threshold for splitting unitigs is 15 minimum ovl 199\n",
      "\u001b[0;32m[Mon Dec  3 21:29:20 UTC 2018]\u001b[0m Running assembly\n",
      "\u001b[0;32m[Mon Dec  3 21:30:09 UTC 2018]\u001b[0m Recomputing A-stat\n",
      "recomputing A-stat for super-reads\n",
      "\u001b[0;32m[Mon Dec  3 21:30:14 UTC 2018]\u001b[0m Mega-reads initial assembly complete\n",
      "\u001b[0;32m[Mon Dec  3 21:30:14 UTC 2018]\u001b[0m No gap closing possible\n",
      "\u001b[0;32m[Mon Dec  3 21:30:14 UTC 2018]\u001b[0m Removing redundant scaffolds\n",
      "\u001b[0;32m[Mon Dec  3 21:30:17 UTC 2018]\u001b[0m Assembly complete, final scaffold sequences are in CA.mr.41.15.15.0.02/final.genome.scf.fasta\n",
      "\u001b[0;32m[Mon Dec  3 21:30:17 UTC 2018]\u001b[0m All done\n",
      "\u001b[0;32m[Mon Dec  3 21:30:17 UTC 2018]\u001b[0m Final stats for CA.mr.41.15.15.0.02/final.genome.scf.fasta\n",
      "N50 1350979\n",
      "Sequence 2046583\n",
      "Average 255823\n",
      "E-size 948826\n",
      "Count 8\n"
     ]
    }
   ],
   "source": [
    "#Config file is loaded and assembly script is run\n",
    "masurca config_file.txt && ./assemble.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "\n",
    "[1] Wick R.R., Judd L.M., Gorrie C.L., Holt K.E. Unicycler: resolving bacterial genome assemblies from short and long sequencing reads. PLoS Comput Biol 2017. DOI: https://doi.org/10.1371/journal.pcbi.1005595\n",
    "\n",
    "[2] Wick R.R., Judd L.M., Gorrie C.L., Holt K.E. Completing bacterial genome assemblies with multiplex MinION sequencing. Microb Genom 2017. DOI: https://dx.doi.org/10.1099/mgen.0.000132\n",
    "\n",
    "[3] Zimin A.V., Marçais G., Puiu D., Roberts M., Salzberg S.S., Yorke J.A. The MaSuRCA genome assembler, Bioinformatics, Volume 29, Issue 21, 1 November 2013, Pages 2669–2677. DOI: https://doi.org/10.1093/bioinformatics/btt476\n",
    "\n",
    "[4] Zimin A.V., Puiu D., Luo M.-C., Zhu T., Koren S., Marcais G., Yorke J.A., Dvorak J., Salzberg S.L. Hybrid assembly of the large and highly repetitive genome of Aegilops tauschii, a progenitor of bread wheat, with the MaSuRCA mega-reads algorithm. DOI: https://doi.org/10.1101/gr.213405.116 \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
